{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "import os\n",
    "import re\n",
    "import shutil\n",
    "import subprocess\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/media/ofornes/Storage/Work/GRECO',\n",
       " '/media/ofornes/Storage/Work/GRECO/JASPAR-profile-inference',\n",
       " '/home/ofornes/Anaconda3/lib/python38.zip',\n",
       " '/home/ofornes/Anaconda3/lib/python3.8',\n",
       " '/home/ofornes/Anaconda3/lib/python3.8/lib-dynload',\n",
       " '',\n",
       " '/home/ofornes/Anaconda3/lib/python3.8/site-packages',\n",
       " '/home/ofornes/Anaconda3/lib/python3.8/site-packages/IPython/extensions',\n",
       " '/home/ofornes/.ipython']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Append JASPAR-profile-inference to path\n",
    "jaspar_dir = os.path.abspath(\"./JASPAR-profile-inference\")\n",
    "sys.path.insert(1, jaspar_dir)\n",
    "sys.path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import from JASPAR-profile-inference\n",
    "from __init__ import Jglobals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------#\n",
    "# Class       #\n",
    "#-------------#\n",
    "\n",
    "class TF(object):\n",
    "\n",
    "    def __init__(self, gene_name, species):\n",
    "\n",
    "        self.gene_name = gene_name\n",
    "        self.species = species\n",
    "        self.uniacc = \"\"\n",
    "        self.unientry = \"\"\n",
    "        self.geneid = \"\"\n",
    "        self.status = \"\"\n",
    "        self.sequence = \"\"\n",
    "        self.family = \"Unknown\"\n",
    "        self.cluster_num = \"\"\n",
    "        # self.orthodb = set()\n",
    "        self.jaspar_id = \"\"\n",
    "        self.hocomoco_id = set()\n",
    "        \n",
    "        # In vivo\n",
    "        self.chip_atlas = set()\n",
    "        self.cistromedb = set()\n",
    "        self.gtrd = set()\n",
    "        self.remap = set()\n",
    "        self.dap_seq = set()\n",
    "\n",
    "        # In vitro\n",
    "        self.ht_selex = set()\n",
    "        self.cisbp = set()\n",
    "        self.uniprobe = set()\n",
    "        self.smile_seq = set()\n",
    "\n",
    "        # Hidden variables (for internal use only)\n",
    "        self._uniaccs = set()\n",
    "        self._unientries = set()\n",
    "        self._geneids = set()\n",
    "        self._sequences = set()\n",
    "\n",
    "    @property\n",
    "    def invivo(self):\n",
    "        \"\"\"\n",
    "        Returns 1 if the TF has been profiled by in vivo methods,\n",
    "        or 0 otherwise.\n",
    "        @rtype = {int}\n",
    "        \"\"\"\n",
    "\n",
    "        # For simplicity, ChIP-seq ~ ChIP-exo ~ DAP-seq\n",
    "        if self.chip_atlas or self.cistromedb or self.gtrd or self.dap_seq or self.remap:\n",
    "            return(1)\n",
    "\n",
    "        return(0)\n",
    "\n",
    "    @property\n",
    "    def invitro(self):\n",
    "        \"\"\"\n",
    "        Returns the number of different experimental methods by\n",
    "        which a TF has been profiled in vitro.\n",
    "        @rtype = {int}\n",
    "        \"\"\"\n",
    "\n",
    "        n = 0\n",
    "\n",
    "        if self.ht_selex:\n",
    "            n += 1\n",
    "        # i.e. PBM\n",
    "        if self.cisbp or self.uniprobe:\n",
    "            n += 1\n",
    "        if self.smile_seq:\n",
    "            n += 1\n",
    "\n",
    "        return(n)\n",
    "\n",
    "    @property\n",
    "    def evidence(self):\n",
    "        \"\"\"\n",
    "        Returns the amount of evidence associated with a TF.\n",
    "        @rtype = {int}\n",
    "        \"\"\"\n",
    "        return(self.invivo+self.invitro)\n",
    "\n",
    "    def __str__(self):\n",
    "\n",
    "        string = \"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\".format(\n",
    "            self.gene_name,\n",
    "            self.geneid,\n",
    "            self.species,\n",
    "            self.uniacc,\n",
    "            # \";\".join(sorted([i for i in self._uniaccs if i != self.uniacc])),\n",
    "            self.unientry,\n",
    "            # \";\".join(sorted([i for i in self._unientries if i != self.unientry])),\n",
    "            self.status,\n",
    "            self.sequence,\n",
    "            # \";\".join(sorted([i for i in self._sequences if i != self.sequence])),\n",
    "            self.family,\n",
    "            self.cluster_num,\n",
    "            self.evidence,\n",
    "            # \";\".join(sorted([i for i in self._pfam_ids if i != self.pfam_id])),\n",
    "            # \";\".join(sorted(self.orthodb)),\n",
    "            self.jaspar_id,\n",
    "            \";\".join(sorted(self.hocomoco_id)),\n",
    "            \";\".join(sorted(self.chip_atlas)),\n",
    "            \";\".join(sorted(self.cistromedb)),\n",
    "            \";\".join(sorted(self.gtrd)),\n",
    "            \";\".join(sorted(self.remap)),\n",
    "            \";\".join(sorted(self.dap_seq)),\n",
    "            \";\".join(sorted(self.ht_selex)),\n",
    "            \";\".join(sorted(self.cisbp)),\n",
    "            \";\".join(sorted(self.uniprobe)),\n",
    "            \";\".join(sorted(self.smile_seq))\n",
    "        )\n",
    "\n",
    "        return(string)\n",
    "\n",
    "    def __repr__(self):\n",
    "\n",
    "        string = \"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\".format(\n",
    "            self.gene_name,\n",
    "            self.geneid,\n",
    "            self.species,\n",
    "            self.uniacc,\n",
    "            # \";\".join(sorted([i for i in self._uniaccs if i != self.uniacc])),\n",
    "            self.unientry,\n",
    "            # \";\".join(sorted([i for i in self._unientries if i != self.unientry])),\n",
    "            self.status,\n",
    "            self.sequence,\n",
    "            # \";\".join(sorted([i for i in self._sequences if i != self.sequence])),\n",
    "            self.family,\n",
    "            self.cluster_num,\n",
    "            self.evidence,\n",
    "            # \";\".join(sorted([i for i in self._pfam_ids if i != self.pfam_id])),\n",
    "            # \";\".join(sorted(self.orthodb)),\n",
    "            self.jaspar_id,\n",
    "            \";\".join(sorted(self.hocomoco_id)),\n",
    "            \";\".join(sorted(self.chip_atlas)),\n",
    "            \";\".join(sorted(self.cistromedb)),\n",
    "            \";\".join(sorted(self.gtrd)),\n",
    "            \";\".join(sorted(self.remap)),\n",
    "            \";\".join(sorted(self.dap_seq)),\n",
    "            \";\".join(sorted(self.ht_selex)),\n",
    "            \";\".join(sorted(self.cisbp)),\n",
    "            \";\".join(sorted(self.uniprobe)),\n",
    "            \";\".join(sorted(self.smile_seq))\n",
    "        )\n",
    "\n",
    "        return(string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------#\n",
    "# Paths       #\n",
    "#-------------#\n",
    "\n",
    "hocomoco_file = \"./Data/Databases/HOCOMOCO/HOCOMOCOv11_full_jaspar_format.txt\"\n",
    "chip_atlas_file = \"./Data/Experiments/ChIP-seq.ChIP-Atlas.tsv\"\n",
    "cistromedb_file = \"./Data/Experiments/ChIP-seq.CistromeDB.tsv\"\n",
    "gtrd_file = \"./Data/Experiments/ChIP-seq.GTRD.tsv\"\n",
    "remap_file = \"./Data/Experiments/ChIP-seq.ReMap2020.tsv\"\n",
    "dap_seq_file = \"./Data/Experiments/DAP-seq.PMID:27203113.tsv\"\n",
    "ht_selex_files = [\"./Data/Experiments/HT-SELEX.PMID:23332764.tsv\",\n",
    "                  \"./Data/Experiments/HT-SELEX.PMID:28473536.tsv\"]\n",
    "cisbp_txt_file = \"./Data/Databases/CisBP-2.0/PWMs.txt\"\n",
    "cisbp_tsv_file = \"./Data/Experiments/PBM.CisBP-2.0.tsv\"\n",
    "uniprobe_file = \"./Data/Experiments/PBM.UniPROBE.tsv\"\n",
    "smile_seq = \"./Data/Experiments/SMiLE-seq.PMID:28092692.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "T26I12.90\t824687\tArabidopsis thaliana\tQ9M3C9\tQ9M3C9_ARATH\tUnreviewed\tMSPPSTIAYVLPPGFKFVPNDEEVIHCYLKPYSDGNTNVLLHVPIHLVNIYESNPQTLSEEFQKGNDKEWFIITERNKVDQGLSQTKRVGYGAKRQKRVDTNGGYWHATVAAQKINAGDGVVRNKRPLAYYVGKPSEGVKTDWLMQEYSLDHSSHNNDKDYTLCKIYLTPQATKMNKEVGEEKKKQKKGEAVVSVAPVEALEEQLPCNVEYHQPLAPLDSCQPQPHDLAYQQQQFCPGPLDSYQPQPHDMENQQPHNEKLKKEEDVEQLDLHQPDQGKGC\tNAC/NAM\t\t0\t\t\t\t\t\t\t\t\t\t\t\n"
     ]
    }
   ],
   "source": [
    "#-------------#\n",
    "# Parse Data  #\n",
    "#-------------#\n",
    "\n",
    "# Initialize\n",
    "tfs = set()\n",
    "\n",
    "# For each line...\n",
    "for line in Jglobals.parse_tsv_file(\"./Data/Parsed/TFs.tab.gz\"):\n",
    "\n",
    "    # Initialize\n",
    "    tf = TF(line[0], line[1])\n",
    "\n",
    "    # Get UniProt Accession\n",
    "    uniaccs = line[2].split(\";\")\n",
    "    tf.uniacc = uniaccs[0]\n",
    "    tf._uniaccs.update(set(uniaccs))\n",
    "\n",
    "    # Get UniProt Entry\n",
    "    unientries = line[3].split(\";\")\n",
    "    tf.unientry = unientries[0]\n",
    "    tf._unientries.update(set(unientries))\n",
    "\n",
    "    # Get Entrez Gene ID\n",
    "    geneids = str(line[4]).split(\";\")\n",
    "    tf.geneid = geneids[0]\n",
    "    tf._geneids.update(set(geneids))\n",
    "\n",
    "    # Get sequence\n",
    "    sequences = line[5].split(\";\")\n",
    "    tf.sequence = sequences[0]\n",
    "    tf._sequences.update(set(sequences))\n",
    "\n",
    "    # Get status (i.e. reviewed or not)\n",
    "    tf.status = line[6]\n",
    "\n",
    "    # Get family\n",
    "    tf.family = line[7]\n",
    "\n",
    "    # Get JASPAR ids\n",
    "    if isinstance(line[8], str):\n",
    "        tf.jaspar_id = line[8]\n",
    "\n",
    "    # Add TF to TFs\n",
    "    tfs.add(tf)\n",
    "\n",
    "print(next(iter(tfs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------#\n",
    "# HOCOMOCO    #\n",
    "#-------------#\n",
    "\n",
    "#   678 Homo sapiens\n",
    "#   451 Mus musculus\n",
    "\n",
    "# SMCA5_MOUSE # Not a TF: SWI/SNF-related matrix-associated actin-dependent regulator of chromatin subfamily A member 5\n",
    "# FUBP1_MOUSE # Not a TF: Far upstream element-binding protein 1\n",
    "# BRCA1_MOUSE # Not a TF: Breast cancer type 1 susceptibility protein homolog\n",
    "# EVI1_MOUSE  # Not a TF: Histone-lysine N-methyltransferase MECOM\n",
    "# TAF1_MOUSE  # Not a TF: Transcription initiation factor TFIID subunit 1\n",
    "# BRAC_MOUSE  # Not a valid UniProt Entry\n",
    "# HLTF_MOUSE  # Not a TF: Helicase-like transcription factor\n",
    "# TAF1_HUMAN  # Not a TF: Transcription initiation factor TFIID subunit 1\n",
    "# HLTF_HUMAN  # Not a TF: Helicase-like transcription factor\n",
    "# BRCA1_HUMAN # Not a TF: Breast cancer type 1 susceptibility protein\n",
    "# SMCA5_HUMAN # Not a TF: SWI/SNF-related matrix-associated actin-dependent regulator of chromatin subfamily A member 5\n",
    "# ZF64A_HUMAN # Not a valid UniProt Entry\n",
    "# BRAC_HUMAN  # Not a valid UniProt Entry\n",
    "# EVI1_HUMAN  # Not a TF: Histone-lysine N-methyltransferase MECOM\n",
    "# FUBP1_HUMAN # Not a TF: Far upstream element-binding protein 1\n",
    "# BPTF_HUMAN  # Not a TF: Nucleosome-remodeling factor subunit BPTF\n",
    "# CENPB_HUMAN # Not a TF: Major centromere autoantigen B\n",
    "# ZBT48_HUMAN # Not a valid UniProt Entry; should be TZAP_HUMAN\n",
    "\n",
    "# For each line...\n",
    "for line in Jglobals.parse_file(hocomoco_file):\n",
    "\n",
    "    if line.startswith(\">\"):\n",
    "\n",
    "        # Get unientry\n",
    "        m = re.search(\"(\\w+_(HUMAN|MOUSE))\", line)\n",
    "        unientry = m.group(1)\n",
    "\n",
    "        # For each TF...\n",
    "        for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "            if unientry in tf._unientries:\n",
    "                tf.hocomoco_id.add(line[1:])\n",
    "                break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AHR {'AHR_HUMAN.H11MO.0.B'}\n",
      "AIRE {'AIRE_HUMAN.H11MO.0.C'}\n",
      "ALX1 {'ALX1_HUMAN.H11MO.0.B'}\n",
      "ALX3 {'ALX3_HUMAN.H11MO.0.D'}\n",
      "ALX3\t257\tHomo sapiens\tO95076\tALX3_HUMAN\tReviewed\tMDPEHCAPFRVGPAPGPYVASGDEPPGPQGTPAAAPHLHPAPPRGPRLTRFPACGPLEPYLPEPAKPPAKYLQDLGPGPALNGGHFYEGPAEAEEKTSKAASFPQLPLDCRGGPRDGPSNLQGSPGPCLASLHLPLSPGLPDSMELAKNKSKKRRNRTTFSTFQLEELEKVFQKTHYPDVYAREQLALRTDLTEARVQVWFQNRRAKWRKRERYGKIQEGRNPFTAAYDISVLPRTDSHPQLQNSLWASPGSGSPGGPCLVSPEGIPSPCMSPYSHPHGSVAGFMGVPAPSAAHPGIYSIHGFPPTLGGHSFEPSSDGDYKSPSLVSLRVKPKEPPGLLNWTT\tHomeodomain\t\t0\tMA0634.1\tALX3_HUMAN.H11MO.0.D\t\t\t\t\t\t\t\t\t\n",
      "...\n",
      "//\n",
      "Total genes: 1057\n",
      "Total feats: 1214\n"
     ]
    }
   ],
   "source": [
    "genes = 0\n",
    "feats = 0\n",
    "count = 0\n",
    "for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "    subtotal_feats = len(tf.hocomoco_id)\n",
    "    if subtotal_feats > 0:\n",
    "        genes += 1\n",
    "        feats += subtotal_feats\n",
    "        if count <= 3:\n",
    "            print(tf.gene_name, tf.hocomoco_id)\n",
    "        if count == 3:\n",
    "            print(tf)\n",
    "            print(\"...\")\n",
    "        else:\n",
    "            pass\n",
    "        count += 1\n",
    "print(\"//\")\n",
    "print(\"Total genes: %s\" % genes)\n",
    "print(\"Total feats: %s\" % feats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------#\n",
    "# ChIP-Atlas  #\n",
    "#-------------#\n",
    "\n",
    "#   155 ce10\n",
    "#   234 dm3\n",
    "#   975 hg19\n",
    "#   723 mm9\n",
    "#    52 rn6\n",
    "#   135 sacCer3\n",
    "\n",
    "genomes = {\n",
    "    \"ce10\": \"Caenorhabditis elegans\",\n",
    "    \"dm3\": \"Drosophila melanogaster\",\n",
    "    \"hg19\": \"Homo sapiens\",\n",
    "    \"mm9\": \"Mus musculus\",\n",
    "    \"rn6\": \"Rattus norvegicus\",\n",
    "    \"sacCer3\": \"Saccharomyces cerevisiae\"\n",
    "}\n",
    "\n",
    "# For each line...\n",
    "for line in Jglobals.parse_tsv_file(chip_atlas_file):\n",
    "\n",
    "    # Inialize\n",
    "    genome = line[0]\n",
    "    antigen_class = line[1]\n",
    "    antigen = line[2]\n",
    "    experiment_ids = line[4]\n",
    "\n",
    "    if antigen_class != \"TFs and others\":\n",
    "        continue\n",
    "\n",
    "    # For each TF...\n",
    "    for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "        if antigen.upper() == tf.gene_name.upper() and genomes[genome] == tf.species:\n",
    "            for experiment_id in experiment_ids.split(\",\"):\n",
    "                tf.chip_atlas.add(experiment_id)\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ABF1 {'SRX3709366', 'SRX3544078', 'SRX3709365'}\n",
      "ADNP2 {'SRX359980'}\n",
      "AEBP2 {'SRX5426124', 'SRX6370210', 'SRX6370192', 'SRX5426125', 'SRX6370209', 'SRX6370191'}\n",
      "AHCTF1 {'SRX359920'}\n",
      "AHCTF1\t25909\tHomo sapiens\tQ8WYP5\tELYS_HUMAN\tReviewed\tMRDLRAQVTSGLLPFPEVTLQALGEDEITLESVLRGKFAAGKNGLACLACGPQLEVVNSITGERLSAYRFSGVNEQPPVVLAVKEFSWQKRTGLLIGLEETEGSVLCLYDLGISKVVKAVVLPGRVTAIEPIINHGGASASTQHLHPSLRWLFGVAAVVTDVGQILLVDLCLDDLSCNQNEVEASDLEVLTGIPAEVPHIRESVMRQGRHLCFQLVSPTGTAVSTLSYISRTNQLAVGFSDGYLALWNMKSMKREYYIQLESGQVPVYAVTFQEPENDPRNCCYLWAVQSTQDSEGDVLSLHLLQLAFGNRKCLASGQILYEGLEYCEERYTLDLTGGMFPLRGQTSNTKLLGCQSIEKFRSHGDREEGVNEALSPDTSVSVFTWQVNIYGQGKPSVYLGLFDINRWYHAQMPDSLRSGEYLHNCSYFALWSLESVVSRTSPHGILDILVHERSLNRGVPPSYPPPEQFFNPSTYNFDATCLLNSGVVHLTCTGFQKETLTFLKKSGPSLNELIPDGYNRCLVAGLLSPRFVDVQPSSLSQEEQLEAILSAAIQTSSLGLLTGYIRRWITEEQPNSATNLRFVLEWTWNKVVLTKEEFDRLCVPLFDGSCHFMDPQTIQSIQQCYLLLSNLNIVLSCFASEAREITERGLIDLSNKFVVSHLICQYAQVVLWFSHSGLLPEGIDDSVQLSRLCYNYPVIQNYYTSRRQKFERLSRGKWNPDCLMIDGLVSQLGERIEKLWKRDEGGTGKYPPASLHAVLDMYLLDGVTEAAKHSITIYLLLDIMYSFPNKTDTPIESFPTVFAISWGQVKLIQGFWLIDHNDYESGLDLLFHPATAKPLSWQHSKIIQAFMSQGEHRQALRYIQTMKPTVSSGNDVILHLTVLLFNRCMVEAWNFLRQHCNRLNIEELLKHMYEVCQEMGLMEDLLKLPFTDTEQECLVKFLQSSASVQNHEFLLVHHLQRANYVPALKLNQTLKINVMNDRDPRLRERSLARNSILDQYGKILPRVHRKLAIERAKPYHLSTSSVFRLVSRPKPLSAVPKQVVTGTVLTRSVFINNVLSKIGEVWASKEPINSTTPFNSSKIEEPSPIVYSLPAPELPEAFFGTPISKASQKISRLLDLVVQPVPRPSQCSEFIQQSSMKSPLYLVSRSLPSSSQLKGSPQAISRASELHLLETPLVVKKAKSLAMSVTTSGFSEFTPQSILRSTLRSTPLASPSPSPGRSPQRLKETRISFVEEDVHPKWIPGAADDSKLEVFTTPKKCAVPVETEWLKSKDRTTSFFLNSPEKEHQEMDEGSQSLEKLDVSKGNSSVSITSDETTLEYQDAPSPEDLEETVFTASKPKSSSTALTTNVTEQTEKDGDKDVFASEVTPSDLQKQMGNLEDAETKDLLVAAEAFSELNHLSPVQGTEASLCAPSVYEGKIFTQKSKVPVLDEGLTSVETYTPAIRANDNKSMADVLGDGGNSSLTISEGPIVSERRLNQEVALNLKEDHEVEVGVLKESVDLPEEKLPISDSPPDTQEIHVIEQEKLEAQDSGEEARNLSFNELYPSGTLKLQYNFDTIDQQFCDLADNKDTAECDIAEVDGELFVAQSNFTLILEGEEGEVEPGDFASSDVLPKAANTATEEKLVCSGENDNHGQIANLPSAVTSDQKSQKVDTLPYVPEPIKVAIAENLLDVIKDTRSKEITSDTMEQSIHETIPLVSQNIMCPTKLVKSAFKTAQETSTMTMNVSQVDDVVSSKTRTRGQRIQNVNVKSAQQEASADVATPKMPGQSVRKKTRKAKEISEASENIYSDVRGLSQNQQIPQNSVTPRRGRRKKEVNQDILENTSSVEQELQITTGRESKRLKSSQLLEPAVEETTKKEVKVSSVTKRTPRRIKRSVENQESVEIINDLKVSTVTSPSRMIRKLRSTNLDASENTGNKQDDKSSDKQLRIKHVRRVRGREVSPSDVREDSNLESSQLTVQAEFDMSAIPRKRGRPRKINPSEDVGSKAVKEERSPKKKEAPSIRRRSTRNTPAKSENVDVGKPALGKSILVPNEELSMVMSSKKKLTKKTESQSQKRSLHSVSEERTDEMTHKETNEQEERLLATASFTKSSRSSRTRSSKAILLPDLSEPNNEPLFSPASEVPRKAKAKKIEVPAQLKELVSDLSSQFVISPPALRSRQKNTSNKNKLEDELKDDAQSVETLGKPKAKRIRTSKTKQASKNTEKESAWSPPPIEIRLISPLASPADGVKSKPRKTTEVTGTGLGRNRKKLSSYPKQILRRKML\tUnknown\t\t1\t\t\tSRX359920\t\t\t\t\t\t\t\t\n",
      "...\n",
      "//\n",
      "Total genes: 1092\n",
      "Total feats: 15775\n"
     ]
    }
   ],
   "source": [
    "genes = 0\n",
    "feats = 0\n",
    "count = 0\n",
    "for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "    subtotal_feats = len(tf.chip_atlas)\n",
    "    if subtotal_feats > 0:\n",
    "        genes += 1\n",
    "        feats += subtotal_feats\n",
    "        if count <= 3:\n",
    "            print(tf.gene_name, tf.chip_atlas)\n",
    "        if count == 3:\n",
    "            print(tf)\n",
    "            print(\"...\")\n",
    "        else:\n",
    "            pass\n",
    "        count += 1\n",
    "print(\"//\")\n",
    "print(\"Total genes: %s\" % genes)\n",
    "print(\"Total feats: %s\" % feats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------#\n",
    "# CistromeDB  #\n",
    "#-------------#\n",
    "\n",
    "# 11348 Homo sapiens\n",
    "#  9060 Mus musculus\n",
    "\n",
    "# For each line...\n",
    "for line in Jglobals.parse_tsv_file(cistromedb_file):\n",
    "\n",
    "    # Inialize\n",
    "    experiment_id = str(line[0])\n",
    "    species = line[1]\n",
    "    gene = line[6]\n",
    "\n",
    "    # For each TF...\n",
    "    for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "        if gene.upper() == tf.gene_name.upper() and species == tf.species:\n",
    "            tf.cistromedb.add(experiment_id)\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ADNP {'63582', '38734', '63581'}\n",
      "ADNP2 {'42808'}\n",
      "AEBP2 {'64454', '64455', '38782'}\n",
      "AHCTF1 {'42748'}\n",
      "AHCTF1\t25909\tHomo sapiens\tQ8WYP5\tELYS_HUMAN\tReviewed\tMRDLRAQVTSGLLPFPEVTLQALGEDEITLESVLRGKFAAGKNGLACLACGPQLEVVNSITGERLSAYRFSGVNEQPPVVLAVKEFSWQKRTGLLIGLEETEGSVLCLYDLGISKVVKAVVLPGRVTAIEPIINHGGASASTQHLHPSLRWLFGVAAVVTDVGQILLVDLCLDDLSCNQNEVEASDLEVLTGIPAEVPHIRESVMRQGRHLCFQLVSPTGTAVSTLSYISRTNQLAVGFSDGYLALWNMKSMKREYYIQLESGQVPVYAVTFQEPENDPRNCCYLWAVQSTQDSEGDVLSLHLLQLAFGNRKCLASGQILYEGLEYCEERYTLDLTGGMFPLRGQTSNTKLLGCQSIEKFRSHGDREEGVNEALSPDTSVSVFTWQVNIYGQGKPSVYLGLFDINRWYHAQMPDSLRSGEYLHNCSYFALWSLESVVSRTSPHGILDILVHERSLNRGVPPSYPPPEQFFNPSTYNFDATCLLNSGVVHLTCTGFQKETLTFLKKSGPSLNELIPDGYNRCLVAGLLSPRFVDVQPSSLSQEEQLEAILSAAIQTSSLGLLTGYIRRWITEEQPNSATNLRFVLEWTWNKVVLTKEEFDRLCVPLFDGSCHFMDPQTIQSIQQCYLLLSNLNIVLSCFASEAREITERGLIDLSNKFVVSHLICQYAQVVLWFSHSGLLPEGIDDSVQLSRLCYNYPVIQNYYTSRRQKFERLSRGKWNPDCLMIDGLVSQLGERIEKLWKRDEGGTGKYPPASLHAVLDMYLLDGVTEAAKHSITIYLLLDIMYSFPNKTDTPIESFPTVFAISWGQVKLIQGFWLIDHNDYESGLDLLFHPATAKPLSWQHSKIIQAFMSQGEHRQALRYIQTMKPTVSSGNDVILHLTVLLFNRCMVEAWNFLRQHCNRLNIEELLKHMYEVCQEMGLMEDLLKLPFTDTEQECLVKFLQSSASVQNHEFLLVHHLQRANYVPALKLNQTLKINVMNDRDPRLRERSLARNSILDQYGKILPRVHRKLAIERAKPYHLSTSSVFRLVSRPKPLSAVPKQVVTGTVLTRSVFINNVLSKIGEVWASKEPINSTTPFNSSKIEEPSPIVYSLPAPELPEAFFGTPISKASQKISRLLDLVVQPVPRPSQCSEFIQQSSMKSPLYLVSRSLPSSSQLKGSPQAISRASELHLLETPLVVKKAKSLAMSVTTSGFSEFTPQSILRSTLRSTPLASPSPSPGRSPQRLKETRISFVEEDVHPKWIPGAADDSKLEVFTTPKKCAVPVETEWLKSKDRTTSFFLNSPEKEHQEMDEGSQSLEKLDVSKGNSSVSITSDETTLEYQDAPSPEDLEETVFTASKPKSSSTALTTNVTEQTEKDGDKDVFASEVTPSDLQKQMGNLEDAETKDLLVAAEAFSELNHLSPVQGTEASLCAPSVYEGKIFTQKSKVPVLDEGLTSVETYTPAIRANDNKSMADVLGDGGNSSLTISEGPIVSERRLNQEVALNLKEDHEVEVGVLKESVDLPEEKLPISDSPPDTQEIHVIEQEKLEAQDSGEEARNLSFNELYPSGTLKLQYNFDTIDQQFCDLADNKDTAECDIAEVDGELFVAQSNFTLILEGEEGEVEPGDFASSDVLPKAANTATEEKLVCSGENDNHGQIANLPSAVTSDQKSQKVDTLPYVPEPIKVAIAENLLDVIKDTRSKEITSDTMEQSIHETIPLVSQNIMCPTKLVKSAFKTAQETSTMTMNVSQVDDVVSSKTRTRGQRIQNVNVKSAQQEASADVATPKMPGQSVRKKTRKAKEISEASENIYSDVRGLSQNQQIPQNSVTPRRGRRKKEVNQDILENTSSVEQELQITTGRESKRLKSSQLLEPAVEETTKKEVKVSSVTKRTPRRIKRSVENQESVEIINDLKVSTVTSPSRMIRKLRSTNLDASENTGNKQDDKSSDKQLRIKHVRRVRGREVSPSDVREDSNLESSQLTVQAEFDMSAIPRKRGRPRKINPSEDVGSKAVKEERSPKKKEAPSIRRRSTRNTPAKSENVDVGKPALGKSILVPNEELSMVMSSKKKLTKKTESQSQKRSLHSVSEERTDEMTHKETNEQEERLLATASFTKSSRSSRTRSSKAILLPDLSEPNNEPLFSPASEVPRKAKAKKIEVPAQLKELVSDLSSQFVISPPALRSRQKNTSNKNKLEDELKDDAQSVETLGKPKAKRIRTSKTKQASKNTEKESAWSPPPIEIRLISPLASPADGVKSKPRKTTEVTGTGLGRNRKKLSSYPKQILRRKML\tUnknown\t\t1\t\t\tSRX359920\t42748\t\t\t\t\t\t\t\n",
      "...\n",
      "//\n",
      "Total genes: 1279\n",
      "Total feats: 11702\n"
     ]
    }
   ],
   "source": [
    "genes = 0\n",
    "feats = 0\n",
    "count = 0\n",
    "for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "    subtotal_feats = len(tf.cistromedb)\n",
    "    if subtotal_feats > 0:\n",
    "        genes += 1\n",
    "        feats += subtotal_feats\n",
    "        if count <= 3:\n",
    "            print(tf.gene_name, tf.cistromedb)\n",
    "        if count == 3:\n",
    "            print(tf)\n",
    "            print(\"...\")\n",
    "        else:\n",
    "            pass\n",
    "        count += 1\n",
    "print(\"//\")\n",
    "print(\"Total genes: %s\" % genes)\n",
    "print(\"Total feats: %s\" % feats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------#\n",
    "# GTRD        #\n",
    "#-------------#\n",
    "\n",
    "#    71 Arabidopsis thaliana\n",
    "#   213 Caenorhabditis elegans\n",
    "#    11 Danio rerio\n",
    "#   249 Drosophila melanogaster\n",
    "#  1236 Homo sapiens\n",
    "#   513 Mus musculus\n",
    "#    12 Rattus norvegicus\n",
    "#   137 Saccharomyces cerevisiae\n",
    "#    32 Schizosaccharomyces pombe\n",
    "\n",
    "# For each line...\n",
    "for line in Jglobals.parse_tsv_file(gtrd_file):\n",
    "\n",
    "    # Inialize\n",
    "    experiment_id = line[0]\n",
    "    uniacc = line[1]\n",
    "\n",
    "    # For each TF...\n",
    "    for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "        if uniacc in tf._uniaccs:\n",
    "            tf.gtrd.add(experiment_id)\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11723 {'EXP043557'}\n",
      "ABF4 {'EXP041418'}\n",
      "ADD1 {'EXP044073'}\n",
      "ADNP {'EXP039553'}\n",
      "ADNP\t23394\tHomo sapiens\tQ9H2P0\tADNP_HUMAN\tReviewed\tMFQLPVNNLGSLRKARKTVKKILSDIGLEYCKEHIEDFKQFEPNDFYLKNTTWEDVGLWDPSLTKNQDYRTKPFCCSACPFSSKFFSAYKSHFRNVHSEDFENRILLNCPYCTFNADKKTLETHIKIFHAPNASAPSSSLSTFKDKNKNDGLKPKQADSVEQAVYYCKKCTYRDPLYEIVRKHIYREHFQHVAAPYIAKAGEKSLNGAVPLGSNAREESSIHCKRCLFMPKSYEALVQHVIEDHERIGYQVTAMIGHTNVVVPRSKPLMLIAPKPQDKKSMGLPPRIGSLASGNVRSLPSQQMVNRLSIPKPNLNSTGVNMMSSVHLQQNNYGVKSVGQGYSVGQSMRLGLGGNAPVSIPQQSQSVKQLLPSGNGRSYGLGSEQRSQAPARYSLQSANASSLSSGQLKSPSLSQSQASRVLGQSSSKPAAAATGPPPGNTSSTQKWKICTICNELFPENVYSVHFEKEHKAEKVPAVANYIMKIHNFTSKCLYCNRYLPTDTLLNHMLIHGLSCPYCRSTFNDVEKMAAHMRMVHIDEEMGPKTDSTLSFDLTLQQGSHTNIHLLVTTYNLRDAPAESVAYHAQNNPPVPPKPQPKVQEKADIPVKSSPQAAVPYKKDVGKTLCPLCFSILKGPISDALAHHLRERHQVIQTVHPVEKKLTYKCIHCLGVYTSNMTASTITLHLVHCRGVGKTQNGQDKTNAPSRLNQSPSLAPVKRTYEQMEFPLLKKRKLDDDSDSPSFFEEKPEEPVVLALDPKGHEDDSYEARKSFLTKYFNKQPYPTRREIEKLAASLWLWKSDIASHFSNKRKKCVRDCEKYKPGVLLGFNMKELNKVKHEMDFDAEWLFENHDEKDSRVNASKTADKKLNLGKEDDSSSDSFENLEEESNESGSPFDPVFEVEPKISNDNPEEHVLKVIPEDASESEEKLDQKEDGSKYETIHLTEEPTKLMHNASDSEVDQDDVVEWKDGASPSESGPGSQQVSDFEDNTCEMKPGTWSDESSQSEDARSSKPAAKKKATMQGDREQLKWKNSSYGKVEGFWSKDQSQWKNASENDERLSNPQIEWQNSTIDSEDGEQFDNMTDGVAEPMHGSLAGVKLSSQQA\tHomeodomain\t\t1\t\t\t\t38734;63581;63582\tEXP039553\t\t\t\t\t\t\n",
      "...\n",
      "//\n",
      "Total genes: 814\n",
      "Total feats: 1958\n"
     ]
    }
   ],
   "source": [
    "genes = 0\n",
    "feats = 0\n",
    "count = 0\n",
    "for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "    subtotal_feats = len(tf.gtrd)\n",
    "    if subtotal_feats > 0:\n",
    "        genes += 1\n",
    "        feats += subtotal_feats\n",
    "        if count <= 3:\n",
    "            print(tf.gene_name, tf.gtrd)\n",
    "        if count == 3:\n",
    "            print(tf)\n",
    "            print(\"...\")\n",
    "        else:\n",
    "            pass\n",
    "        count += 1\n",
    "print(\"//\")\n",
    "print(\"Total genes: %s\" % genes)\n",
    "print(\"Total feats: %s\" % feats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------#\n",
    "# ReMap 2020  #\n",
    "#-------------#\n",
    "\n",
    "#   405 Arabidopsis thaliana\n",
    "#  1135 Homo sapiens\n",
    "\n",
    "# For each line...\n",
    "for line in Jglobals.parse_tsv_file(remap_file):\n",
    "\n",
    "    # Inialize\n",
    "    gene = line[0]\n",
    "    species = line[1]\n",
    "    experiment_ids = line[2]\n",
    "\n",
    "    # For each TF...\n",
    "    for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "        if gene.upper() == tf.gene_name.upper() and species == tf.species:\n",
    "            for experiment_id in experiment_ids.split(\";\"):\n",
    "                tf.remap.add(experiment_id)\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ABF1 {'GSE80564.ABF1.Col-0_seedling_3d-EtOH', 'GSE80564.ABF1.Col-0_seedling_3d-ABA'}\n",
      "ABF3 {'GSE80564.ABF3.Col-0_seedling_3d-ABA', 'GSE80564.ABF3.Col-0_seedling_3d-EtOH'}\n",
      "ABF4 {'GSE80564.ABF4.Col-0_seedling_3d-EtOH', 'GSE80564.ABF4.Col-0_seedling_3d-ABA'}\n",
      "ABI5 {'GSE60141.ABI5.Col-0_leaves_tnt_col', 'GSE60142.ABI5.Col-0_seedling_36h-ypet', 'GSE60142.ABI5.Col-0_seedling_36h-wt'}\n",
      "ABI5\t818199.0\tArabidopsis thaliana\tQ9SJN0\tABI5_ARATH\tReviewed\tMVTRETKLTSEREVESSMAQARHNGGGGGENHPFTSLGRQSSIYSLTLDEFQHALCENGKNFGSMNMDEFLVSIWNAEENNNNQQQAAAAAGSHSVPANHNGFNNNNNNGGEGGVGVFSGGSRGNEDANNKRGIANESSLPRQGSLTLPAPLCRKTVDEVWSEIHRGGGSGNGGDSNGRSSSSNGQNNAQNGGETAARQPTFGEMTLEDFLVKAGVVREHPTNPKPNPNPNQNQNPSSVIPAAAQQQLYGVFQGTGDPSFPGQAMGVGDPSGYAKRTGGGGYQQAPPVQAGVCYGGGVGFGAGGQQMGMVGPLSPVSSDGLGHGQVDNIGGQYGVDMGGLRGRKRVVDGPVEKVVERRQRRMIKNRESAARSRARKQAYTVELEAELNQLKEENAQLKHALAELERKRKQQYFESLKSRAQPKLPKSNGRLRTLMRNPSCPL\tbZIP\t\t1\tMA0931.1\t\t\t\t\tGSE60141.ABI5.Col-0_leaves_tnt_col;GSE60142.ABI5.Col-0_seedling_36h-wt;GSE60142.ABI5.Col-0_seedling_36h-ypet\t\t\t\t\t\n",
      "...\n",
      "//\n",
      "Total genes: 1093\n",
      "Total feats: 4691\n"
     ]
    }
   ],
   "source": [
    "genes = 0\n",
    "feats = 0\n",
    "count = 0\n",
    "for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "    subtotal_feats = len(tf.remap)\n",
    "    if subtotal_feats > 0:\n",
    "        genes += 1\n",
    "        feats += subtotal_feats\n",
    "        if count <= 3:\n",
    "            print(tf.gene_name, tf.remap)\n",
    "        if count == 3:\n",
    "            print(tf)\n",
    "            print(\"...\")\n",
    "        else:\n",
    "            pass\n",
    "        count += 1\n",
    "print(\"//\")\n",
    "print(\"Total genes: %s\" % genes)\n",
    "print(\"Total feats: %s\" % feats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------#\n",
    "# DAP-seq     #\n",
    "#-------------#\n",
    "\n",
    "#   934 Arabidopsis thaliana\n",
    "#     2 Zea mays\n",
    "\n",
    "# For each line...\n",
    "for line in Jglobals.parse_tsv_file(dap_seq_file):\n",
    "\n",
    "    if line[0] == \"AvgSpotLen\":\n",
    "        continue\n",
    "\n",
    "    species = line[8]\n",
    "    sra_run = line[9]\n",
    "    gene = line[15]\n",
    "\n",
    "    # For each TF...\n",
    "    for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "        if gene.upper() == tf.gene_name.upper() and species == tf.species:\n",
    "            tf.dap_seq.add(sra_run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ABF2 {'SRR2926839'}\n",
      "ABI5 {'SRR2926840', 'SRR2926841'}\n",
      "ABR1 {'SRR2926082', 'SRR2926081'}\n",
      "AGL13 {'SRR2926463'}\n",
      "AGL13\t825284.0\tArabidopsis thaliana\tQ38837\tAGL13_ARATH\tReviewed\tMGRGKVEVKRIENKITRQVTFSKRKSGLLKKAYELSVLCDAEVSLIIFSTGGKLYEFSNVGVGRTIERYYRCKDNLLDNDTLEDTQGLRQEVTKLKCKYESLLRTHRNLVGEDLEGMSIKELQTLERQLEGALSATRKQKTQVMMEQMEELRRKERELGDINNKLKLETEDHDFKGFQDLLLNPVLTAGCSTDFSLQSTHQNYISDCNLGYFLQIGFQQHYEQGEGSSVTKSNARSDAETNFVQ\tMADS box\t\t1\tMA1204.1\t\t\t\t\t\tSRR2926463\t\t\t\t\n",
      "...\n",
      "//\n",
      "Total genes: 290\n",
      "Total feats: 487\n"
     ]
    }
   ],
   "source": [
    "genes = 0\n",
    "feats = 0\n",
    "count = 0\n",
    "for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "    subtotal_feats = len(tf.dap_seq)\n",
    "    if subtotal_feats > 0:\n",
    "        genes += 1\n",
    "        feats += subtotal_feats\n",
    "        if count <= 3:\n",
    "            print(tf.gene_name, tf.dap_seq)\n",
    "        if count == 3:\n",
    "            print(tf)\n",
    "            print(\"...\")\n",
    "        else:\n",
    "            pass\n",
    "        count += 1\n",
    "print(\"//\")\n",
    "print(\"Total genes: %s\" % genes)\n",
    "print(\"Total feats: %s\" % feats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------#\n",
    "# HT-SELEX    #\n",
    "#-------------#\n",
    "\n",
    "# For each file...\n",
    "for file_name in ht_selex_files:\n",
    "\n",
    "    m = re.search(\"HT-SELEX.PMID:(\\d+).tsv\", file_name)\n",
    "    pmid = int(m.group(1))\n",
    "\n",
    "    # For each line...\n",
    "    for line in Jglobals.parse_tsv_file(file_name):\n",
    "\n",
    "        if line[0] == \"Alias\":\n",
    "            continue\n",
    "\n",
    "        m = re.search(\"^([A-Za-z\\d]+)_\", line[0])\n",
    "        if m:\n",
    "\n",
    "            gene_name = m.group(1)\n",
    "\n",
    "            if pmid == 23332764:\n",
    "                sra_run = line[14]\n",
    "            else:\n",
    "                sra_run = line[19]\n",
    "\n",
    "            # For each TF...\n",
    "            for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "                # HT-SELEX data only available for human and mouse\n",
    "                if tf.species not in [\"Homo sapiens\", \"Mus musculus\"]:\n",
    "                    continue\n",
    "                if tf.gene_name == gene_name:\n",
    "                    tf.ht_selex.add(sra_run)\n",
    "                    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ALX1 {'ERR1002054', 'ERR1002055', 'ERR1002057', 'ERR1002056', 'ERR1002050', 'ERR1002053', 'ERR1002052', 'ERR1002051'}\n",
      "ALX3 {'ERR1002063', 'ERR1002067', 'ERR195525', 'ERR1002061', 'ERR1002072', 'ERR195526', 'ERR1002064', 'ERR193772', 'ERR1011038', 'ERR1002068', 'ERR1002069', 'ERR1002062', 'ERR1002065', 'ERR193770', 'ERR193773', 'ERR193771', 'ERR1010299', 'ERR1002071', 'ERR1002073', 'ERR1002070', 'ERR1002058', 'ERR1002060', 'ERR1002066', 'ERR195527', 'ERR1011404', 'ERR195528', 'ERR1010665', 'ERR1002059'}\n",
      "ALX4 {'ERR194918', 'ERR1002081', 'ERR1002075', 'ERR1002074', 'ERR1002076', 'ERR194919', 'ERR1002079', 'ERR1002078', 'ERR1002080', 'ERR194917', 'ERR1002077', 'ERR194920'}\n",
      "AR {'ERR193680', 'ERR193971', 'ERR193973', 'ERR193681', 'ERR193972', 'ERR193970', 'ERR193679', 'ERR193678'}\n",
      "AR\t367\tHomo sapiens\tP10275\tANDR_HUMAN\tReviewed\tMEVQLGLGRVYPRPPSKTYRGAFQNLFQSVREVIQNPGPRHPEAASAAPPGASLLLLQQQQQQQQQQQQQQQQQQQQQQQETSPRQQQQQQGEDGSPQAHRRGPTGYLVLDEEQQPSQPQSALECHPERGCVPEPGAAVAASKGLPQQLPAPPDEDDSAAPSTLSLLGPTFPGLSSCSADLKDILSEASTMQLLQQQQQEAVSEGSSSGRAREASGAPTSSKDNYLGGTSTISDNAKELCKAVSVSMGLGVEALEHLSPGEQLRGDCMYAPLLGVPPAVRPTPCAPLAECKGSLLDDSAGKSTEDTAEYSPFKGGYTKGLEGESLGCSGSAAAGSSGTLELPSTLSLYKSGALDEAAAYQSRDYYNFPLALAGPPPPPPPPHPHARIKLENPLDYGSAWAAAAAQCRYGDLASLHGAGAAGPGSGSPSAAASSSWHTLFTAEEGQLYGPCGGGGGGGGGGGGGGGGGGGGGGGEAGAVAPYGYTRPPQGLAGQESDFTAPDVWYPGGMVSRVPYPSPTCVKSEMGPWMDSYSGPYGDMRLETARDHVLPIDYYFPPQKTCLICGDEASGCHYGALTCGSCKVFFKRAAEGKQKYLCASRNDCTIDKFRRKNCPSCRLRKCYEAGMTLGARKLKKLGNLKLQEEGEASSTTSPTEETTQKLTVSHIEGYECQPIFLNVLEAIEPGVVCAGHDNNQPDSFAALLSSLNELGERQLVHVVKWAKALPGFRNLHVDDQMAVIQYSWMGLMVFAMGWRSFTNVNSRMLYFAPDLVFNEYRMHKSRMYSQCVRMRHLSQEFGWLQITPQEFLCMKALLLFSIIPVDGLKNQKFFDELRMNYIKELDRIIACKRKNPTSCSRRFYQLTKLLDSVQPIARELHQFTFDLLIKSHMVSVDFPEMMAEIISVQVPKILSGKVKPIYFHTQ\tNuclear receptor\t\t2\tMA0007.2\tANDR_HUMAN.H11MO.0.A;ANDR_HUMAN.H11MO.1.A;ANDR_HUMAN.H11MO.2.A\tERX2049941;ERX2049942;ERX2049943;SRX022562;SRX022564;SRX022566;SRX022567;SRX022568;SRX022569;SRX022570;SRX022571;SRX022572;SRX022577;SRX059703;SRX059704;SRX059705;SRX083216;SRX083217;SRX083218;SRX083219;SRX092447;SRX092448;SRX099182;SRX099183;SRX101130;SRX101131;SRX101132;SRX101133;SRX101134;SRX101135;SRX101136;SRX101137;SRX1033266;SRX1033267;SRX1033268;SRX1033269;SRX1033275;SRX1033278;SRX1033281;SRX1033284;SRX1033288;SRX1033289;SRX1067066;SRX1067067;SRX1067068;SRX1067069;SRX1067070;SRX1067071;SRX1070160;SRX1070161;SRX1070162;SRX1070163;SRX1132445;SRX1132446;SRX1132447;SRX1132448;SRX1183808;SRX1183809;SRX1183812;SRX1183813;SRX1334491;SRX1334492;SRX1334493;SRX1334494;SRX1339015;SRX1339016;SRX1339017;SRX1339018;SRX1339019;SRX1339020;SRX142906;SRX142907;SRX147604;SRX147605;SRX1569767;SRX1569768;SRX1569781;SRX1569782;SRX160734;SRX1629233;SRX1629234;SRX1629235;SRX1629236;SRX1629237;SRX1629238;SRX1629239;SRX1629240;SRX1629241;SRX1629242;SRX1629243;SRX1629244;SRX1629245;SRX1629246;SRX1629247;SRX1629248;SRX1629249;SRX1629250;SRX1640090;SRX1640091;SRX1702478;SRX1702479;SRX1702480;SRX1703573;SRX1703574;SRX173191;SRX173192;SRX173193;SRX173194;SRX173195;SRX173196;SRX173201;SRX173202;SRX173206;SRX173207;SRX176048;SRX176049;SRX176050;SRX176051;SRX176053;SRX176054;SRX176055;SRX176056;SRX176057;SRX176058;SRX1818072;SRX1818073;SRX1818074;SRX1818075;SRX1818076;SRX1818077;SRX1885173;SRX1885174;SRX1885175;SRX1885176;SRX1885177;SRX1885178;SRX1885179;SRX1885180;SRX1950170;SRX1950171;SRX1950172;SRX1950173;SRX2017926;SRX2017927;SRX2017936;SRX2017937;SRX2140286;SRX2140287;SRX2140289;SRX2140291;SRX2140294;SRX2140296;SRX2140297;SRX2140299;SRX2140301;SRX2140302;SRX219988;SRX219989;SRX220000;SRX220001;SRX220086;SRX220087;SRX220088;SRX220089;SRX220090;SRX220091;SRX220092;SRX220093;SRX220094;SRX2353645;SRX2353647;SRX2353649;SRX2353651;SRX2353653;SRX2353655;SRX2353657;SRX2353659;SRX2388755;SRX2388756;SRX2388757;SRX2420128;SRX2420129;SRX2420131;SRX2420132;SRX2420134;SRX2420135;SRX250092;SRX2541222;SRX2541227;SRX2541229;SRX2541242;SRX2541243;SRX2541246;SRX2541247;SRX2545037;SRX2545038;SRX2545039;SRX2545040;SRX2633442;SRX2633443;SRX2633444;SRX2633446;SRX2633447;SRX2642351;SRX2642352;SRX2737719;SRX283756;SRX283757;SRX3032991;SRX3032993;SRX3032995;SRX3032997;SRX3055410;SRX3055413;SRX306506;SRX306507;SRX306508;SRX306514;SRX306515;SRX306516;SRX306517;SRX306518;SRX306519;SRX3068918;SRX3068919;SRX3068922;SRX3068923;SRX315145;SRX315146;SRX315147;SRX315148;SRX315149;SRX315150;SRX315151;SRX315152;SRX315153;SRX315155;SRX315156;SRX3229133;SRX3229134;SRX3229166;SRX3229167;SRX3229168;SRX3229169;SRX3229170;SRX3229171;SRX3229172;SRX3229173;SRX3229174;SRX334236;SRX334237;SRX334240;SRX334241;SRX334244;SRX334245;SRX334248;SRX334249;SRX334252;SRX334253;SRX334256;SRX334257;SRX334260;SRX334261;SRX334264;SRX3366971;SRX3366972;SRX3366973;SRX3366974;SRX3366975;SRX3366976;SRX3366983;SRX3366984;SRX3366987;SRX3366988;SRX3366989;SRX3366990;SRX3366995;SRX3366996;SRX3366999;SRX3367000;SRX3367003;SRX3367004;SRX3421794;SRX3474300;SRX3474301;SRX3474302;SRX357292;SRX357293;SRX357296;SRX357297;SRX3584192;SRX3584193;SRX3584194;SRX3584195;SRX365799;SRX383815;SRX383816;SRX3881550;SRX4062879;SRX4062880;SRX4062881;SRX4062882;SRX4108913;SRX4108914;SRX4108915;SRX4108916;SRX4115983;SRX4115984;SRX4116000;SRX4116001;SRX4116006;SRX4116007;SRX4116012;SRX4116013;SRX4116019;SRX4116020;SRX4116070;SRX4116073;SRX4287638;SRX4287639;SRX433189;SRX433190;SRX433191;SRX433192;SRX433193;SRX433194;SRX433196;SRX433197;SRX433198;SRX433200;SRX433201;SRX433202;SRX4411662;SRX4411663;SRX4411672;SRX4411673;SRX4411674;SRX4411675;SRX4582746;SRX4582747;SRX4582748;SRX4582749;SRX470039;SRX471204;SRX471205;SRX471207;SRX471208;SRX471847;SRX471848;SRX471849;SRX471850;SRX471851;SRX471852;SRX471853;SRX471854;SRX471855;SRX471856;SRX4779733;SRX4779735;SRX4783021;SRX4783022;SRX4783023;SRX4783024;SRX4783025;SRX4783026;SRX4783027;SRX4783028;SRX4783029;SRX4783030;SRX4783031;SRX4783032;SRX4783033;SRX4783034;SRX4783035;SRX4783036;SRX4783037;SRX4783038;SRX4783039;SRX4783040;SRX4783041;SRX4783042;SRX4783043;SRX4783044;SRX4783045;SRX4783046;SRX4783047;SRX4783048;SRX4783049;SRX4783050;SRX4783051;SRX4783052;SRX4783053;SRX4783054;SRX4783055;SRX4783056;SRX4783057;SRX4783058;SRX4783059;SRX4783060;SRX4783061;SRX4783062;SRX4783063;SRX4783064;SRX4783065;SRX4783066;SRX4783067;SRX4783068;SRX4783069;SRX4783070;SRX4783071;SRX4783072;SRX4783073;SRX4783074;SRX4783075;SRX4783076;SRX4783077;SRX4783078;SRX4783079;SRX4783080;SRX4783081;SRX4783082;SRX4783083;SRX4783084;SRX4783085;SRX4783086;SRX4783087;SRX4783088;SRX4783089;SRX4783090;SRX4783091;SRX4783092;SRX4783093;SRX4783094;SRX4783095;SRX4783096;SRX4783097;SRX4783098;SRX4783099;SRX4783100;SRX4783101;SRX4783102;SRX4783103;SRX4783104;SRX4783105;SRX4783106;SRX4783107;SRX4783108;SRX4824359;SRX4824360;SRX4824361;SRX4824362;SRX4824363;SRX4824364;SRX4824365;SRX4824366;SRX497604;SRX497605;SRX502069;SRX502070;SRX502071;SRX502072;SRX502073;SRX5124835;SRX5124838;SRX5124841;SRX5124844;SRX5124855;SRX5124857;SRX5124859;SRX5124865;SRX5124867;SRX5124869;SRX5286305;SRX5286306;SRX5286307;SRX5286308;SRX5286309;SRX5286310;SRX5316103;SRX5316110;SRX5401652;SRX5401653;SRX5401656;SRX5401657;SRX5401658;SRX5401659;SRX5401662;SRX5401663;SRX5493746;SRX5493747;SRX5493748;SRX5493749;SRX5493750;SRX5493751;SRX5493752;SRX5493753;SRX5493754;SRX5577150;SRX5577151;SRX5577152;SRX5577153;SRX5577154;SRX5577155;SRX5577156;SRX5577157;SRX5577158;SRX5577159;SRX5577160;SRX5577161;SRX5577162;SRX5577163;SRX5577164;SRX5577165;SRX5577166;SRX5577167;SRX5852998;SRX5852999;SRX5853000;SRX5853001;SRX595671;SRX595675;SRX595676;SRX595677;SRX595685;SRX595687;SRX595688;SRX595690;SRX595693;SRX596536;SRX596537;SRX596538;SRX667982;SRX667983;SRX6878591;SRX6878592;SRX6878593;SRX6878594;SRX698190;SRX698191;SRX698192;SRX698193;SRX698194;SRX698195;SRX698196;SRX698197;SRX734409;SRX734410;SRX734411;SRX734412;SRX736053;SRX736054;SRX736065;SRX736066;SRX736074;SRX736075;SRX758011;SRX847028;SRX847029;SRX847030;SRX847031;SRX847032;SRX847033;SRX847034;SRX847035;SRX861678;SRX861679;SRX861680;SRX861681;SRX861682;SRX861683;SRX861684;SRX861685;SRX882952;SRX882954;SRX882958;SRX882962\t2171;2172;2173;2174;2176;2181;2182;2183;2910;2911;2912;2913;2914;2986;33920;33921;34330;35764;35765;35766;35771;35772;35775;35777;35778;35779;35780;36845;37089;37092;37094;37096;37098;37099;37103;37104;37106;37107;38789;39378;39389;40252;40254;40344;40345;40346;40347;40349;40350;40352;40353;40354;40356;40358;43395;43396;43853;43854;43855;43856;43857;43858;43859;43860;44310;44311;44314;44315;44318;44319;44322;44323;44326;44327;44330;44331;44334;44335;44338;44574;44575;44576;44577;44578;44579;44580;44581;44582;44583;47323;47331;47333;47335;48084;48085;48086;48087;48088;48089;48091;48092;48093;48095;48096;48097;48182;48183;49158;49160;50041;50042;50043;50044;50045;50046;50047;50048;50049;50050;50051;50052;50053;50054;50055;50056;50057;50058;50059;50060;50100;50102;50254;50258;50259;50260;50268;50270;50271;50273;50276;51048;51049;51267;51268;51736;51737;51748;51749;51757;51758;51837;52249;52250;52251;52537;52538;52539;52540;52541;52542;52543;52544;52545;55162;55271;55273;55277;55281;55751;55752;56106;56107;56108;56109;56110;56111;56112;56113;56114;56115;56116;56117;56118;56119;56120;56121;56122;56123;56124;56125;56204;56205;56206;56207;56208;56209;56210;56211;56254;56255;56256;56257;56264;56265;56267;56388;56389;56390;56391;56392;56393;56394;56395;5666;56669;5667;56670;57275;57276;57277;57278;57279;57280;67733;67736;67739;67742;67745;67751;67752;67753;67754;68073;68074;68075;68076;68077;68078;68079;68080;68089;68090;68091;68092;68217;68218;68239;68240;68398;68399;68400;68401;69273;69274;69275;69276;69277;69278;69279;69280;69281;69282;69283;69284;69285;69286;69287;69288;69289;69290;70140;70141;70142;70143;71107;71108;71109;71110;71166;71167;71168;71169;71170;71171;71220;71221;71222;71223;71224;71225;71226;71227;72970;72971;72974;72975;74410;74411;74839;74840;76153;76154;76155;76156;76204;76205;76206;81492;81558;81774;82536;82537;83103;83105;83108;83142;83143;83167;83169;84;84008;84184;84224;84225;85612;85613;85614;85616;85617;85854;85855;85856;85857;85858;86007;8651;8652;8653;86730;86735;87051;87052;87789;87790;87791;87792;87793;87794;87795;87796;87797;88016;88017;88049;88282;88283;88420;88421;88423;9261;9262\tEXP036887;EXP037067;EXP037263;EXP038106;EXP038188;EXP038189;EXP038191;EXP038192;EXP038193;EXP038636;EXP040660;EXP040661;EXP040665\tERP001226.AR.LNCaP;ERP001226.AR.MCF-7;ERP001226.AR.MDA-MB-453;ERP003272.AR.MDA-MB-453_SHFOXA1;ERP003503.AR.LNCaP;ERP003503.AR.LNCaP_FOXA1;ERP003503.AR.MDA-MB-453;ERP003503.AR.MDA-MB-453_FOXA1;GSE104399.AR.MCF-7;GSE104399.AR.breast_tumor_Female_8;GSE104399.AR.breast_tumor_Male_10;GSE104399.AR.breast_tumor_Male_17;GSE104399.AR.breast_tumor_Male_18;GSE104399.AR.breast_tumor_Male_20;GSE104399.AR.breast_tumor_Male_26;GSE104399.AR.breast_tumor_Male_28;GSE104399.AR.breast_tumor_Male_30;GSE104399.AR.breast_tumor_Male_7;GSE104399.AR.breast_tumor_Male_8;GSE109061.AR.epididymis_HEE;GSE109061.AR.epididymis_HEE_R1881;GSE110655.AR.LNCaP;GSE110655.AR.LNCaP_ERG;GSE110655.AR.VCaP_shCt;GSE110655.AR.VCaP_shERG;GSE116189.AR.A-375;GSE116189.AR.A-375_SLNCR;GSE28950.AR.VCaP;GSE28950.AR.VCaP_DHAT_18H;GSE28950.AR.VCaP_DHAT_2H;GSE30623.AR.LNCaP_1F5;GSE30623.AR.LNCaP_1F5_SIFOXA1;GSE31294.AR.LNCaP;GSE31294.AR.LNCaP_KDP53;GSE32892.AR.VCaP;GSE32892.AR.VCaP_R1881;GSE32892.AR.VCaP_R1881_10C26;GSE32892.AR.VCaP_R1881_10C30;GSE32892.AR.VCaP_R1881_1C26;GSE32892.AR.VCaP_R1881_1C30;GSE37345.AR.LNCaP_SHCTR_R1881;GSE37345.AR.LNCaP_SHCTR_R1881_HD;GSE37345.AR.LNCaP_SHFOXA1_R1881;GSE37345.AR.LNCaP_SHFOXA1_R1881_HD;GSE39459.AR.LNCaP-abl;GSE40050.AR.22Rv1;GSE40050.AR.22Rv1_DHT;GSE40050.AR.LNCaP-C4-2B;GSE40050.AR.LNCaP-C4-2B_DHT;GSE40050.AR.LNCaP_DHT;GSE43720.AR.LNCaP;GSE43720.AR.LNCaP_DHT;GSE47120.AR.LNCaP_SHCTR;GSE47120.AR.LNCaP_SHETV1;GSE47987.AR.DU145;GSE47987.AR.DU145_ARC562S;GSE47987.AR.DU145_ARG56W;GSE47987.AR.DU145_ARQ6540X;GSE47987.AR.DU145_FOXA1;GSE47987.AR.DU145_FOXA1_ARC562S;GSE47987.AR.DU145_FOXA1_ARG56W;GSE47987.AR.DU145_FOXA1_ARQ6540X;GSE47987.AR.DU145_FOXA1_HIGH;GSE48930.AR.MCF-7;GSE52725.AR.LNCaP_SHCTR;GSE52725.AR.LNCaP_SHGATA2;GSE54110.AR.PC-3;GSE54110.AR.PC-3_R1881;GSE56086.AR.VCaP;GSE56086.AR.VCaP_SHPIAS1_R1881;GSE56288.AR.LHSAR_HOXB13;GSE56288.AR.LHSAR_LACZ;GSE56288.AR.PLHSAR_FOXA1_HOXB13;GSE56288.AR.prostate;GSE56288.AR.prostate_cancer;GSE58428.AR.LNCaP_DHT24H;GSE58428.AR.LNCaP_DHT24H_SHFOXA1;GSE58428.AR.VCaP_DHT24H;GSE58428.AR.VCaP_DHT24H_SHFOXA1;GSE58428.AR.VCaP_DHT24H_SHFOXP1;GSE58428.AR.VCaP_DHT24H_SHRUNX1;GSE58428.AR.VCaP_ETOH24H;GSE61268.AR.LNCaP-C4-2B_ETOH_SHCTR;GSE61268.AR.LNCaP-C4-2B_R1881_SHCTR;GSE61268.AR.LNCaP_ETOH_SHCTR;GSE61268.AR.LNCaP_HOTAIR;GSE61268.AR.LNCaP_R1881;GSE61268.AR.LNCaP_R1881_HOTAIR;GSE61268.AR.LNCaP_R1881_SHCTR;GSE61838.AR.prostate_DHT;GSE61838.AR.prostate_ETOH;GSE62492.AR.LNCaP;GSE62492.AR.LNCaP_R1881;GSE62492.AR.LNCaP_SHCTR_DHT;GSE62492.AR.LNCaP_SHFOXP1_DHT;GSE63202.AR.LNCaP;GSE64656.AR.LNCaP;GSE65478.AR.prostate;GSE69043.AR.LNCaP_ETOH;GSE69043.AR.LNCaP_PHFRPMIFCS;GSE69043.AR.LNCaP_R1881;GSE69043.AR.LNCaP_RPMIFBS;GSE69043.AR.LNCaP_SHFOXA1_PHFRPMIFCS;GSE69043.AR.LNCaP_SHFOXA1_RPMIFBS;GSE69043.AR.LNCaP_SHGATA2_ETOH;GSE69043.AR.LNCaP_SHGATA2_R1881;GSE70161.AR.MDA-MB-453_R1881;GSE70161.AR.MDA-MB-453_R1881_SICTR;GSE70161.AR.MDA-MB-453_R1881_SIPIAS1;GSE70679.AR.DUCAP;GSE70679.AR.DUCAP_ANDROGEN;GSE72714.AR.LNCaP-C4-2B;GSE72714.AR.LNCaP-C4-2B_SR2211;GSE74069.AR.MDA-MB-453;GSE74069.AR.MDA-MB-453_DHT;GSE74069.AR.MDA-MB-453_MPA;GSE79128.AR.VCaP_DHT;GSE79128.AR.VCaP_R1881;GSE79128.AR.VCaP_SH1_DHT;GSE79128.AR.VCaP_SH1_R1881;GSE79128.AR.VCaP_SH2_DHT;GSE79128.AR.VCaP_SH2_R1881;GSE79128.AR.VCaP_SH3_DHT;GSE79357.AR.LNCaP;GSE79357.AR.LNCaP_NOV;GSE80238.AR.LNCaP-abl_DMSO;GSE80256.AR.LNCaP;GSE80742.AR.22Rv1_R1881;GSE80742.AR.22Rv1_siARFL;GSE80742.AR.22Rv1_siARFL_R1881;GSE83860.AR.LNCaP_DHT;GSE83860.AR.LNCaP_DHT_TNFA;GSE83860.AR.LNCaP_DMSO;GSE84432.AR.VCaP_R1881_30M;GSE84432.AR.VCaP_R1881_4H;GSE85558.AR.22Rv1;GSE85558.AR.22Rv1_Dox;GSE85558.AR.LNCaP;GSE85558.AR.LNCaP_HNF4G_ovexp;GSE86454.AR.NSC_DHT_10nM;GSE86456.AR.LNCaP_3279;GSE86456.AR.LNCaP_DHT_100nM_3279;GSE86456.AR.LNCaP_DHT_100nM_N20;GSE86456.AR.LNCaP_N20;GSE89938.AR.LNCaP_Bag-1L_KO_DHT;GSE89938.AR.LNCaP_Bag-1L_KO_Veh;GSE89938.AR.LNCaP_Bag-1L_WT;GSE89938.AR.LNCaP_Bag-1L_WT_DHT;GSE89938.AR.LNCaP_Bag1L_KO_Bag1L_Rescue_DHT_4h;GSE89938.AR.LNCaP_DHT_Bag-1L-CMut;GSE89938.AR.LNCaP_Talen_DHT;GSE89938.AR.LNCaP_Talen_Veh;GSE90772.AR.myofibroblast;GSE90772.AR.prostate_cancer-associated-fibroblasts;GSE90772.AR.prostate_fetal-fibroblast;GSE92347.AR.LNCaP_DHT;GSE92347.AR.LNCaP_SRC_DHT;GSE92347.AR.VCaP;GSE92347.AR.VCaP_DHT;GSE94577.AR.LNCaP_DHT-1nM;GSE94577.AR.LTAD_DHT-1nM;GSE94577.AR.LTAD_EtOH;GSE94577.AR.LTAD_siCOBLL1;GSE94577.AR.LTAD_siControl;GSE94577.AR.VCaP-LTAD_DHT_10nM;GSE94577.AR.VCaP-LTAD_DHT_1nM;GSE94682.AR.LNCaP;GSE94682.AR.LNCaP_r1881;GSE96652.AR.22Rv1;GSE97831.AR.LNCaP-C4-2;GSE98809.AR.VCaP\t\tERR193678;ERR193679;ERR193680;ERR193681;ERR193970;ERR193971;ERR193972;ERR193973\t\t\t\n",
      "...\n",
      "//\n",
      "Total genes: 672\n",
      "Total feats: 9450\n"
     ]
    }
   ],
   "source": [
    "genes = 0\n",
    "feats = 0\n",
    "count = 0\n",
    "for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "    subtotal_feats = len(tf.ht_selex)\n",
    "    if subtotal_feats > 0:\n",
    "        genes += 1\n",
    "        feats += subtotal_feats\n",
    "        if count <= 3:\n",
    "            print(tf.gene_name, tf.ht_selex)\n",
    "        if count == 3:\n",
    "            print(tf)\n",
    "            print(\"...\")\n",
    "        else:\n",
    "            pass\n",
    "        count += 1\n",
    "print(\"//\")\n",
    "print(\"Total genes: %s\" % genes)\n",
    "print(\"Total feats: %s\" % feats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------#\n",
    "# CIS-BP      #\n",
    "#-------------#\n",
    "\n",
    "valid_matrix_ids = set()\n",
    "\n",
    "# Get valid PWMs\n",
    "for matrix_id in Jglobals.parse_file(cisbp_txt_file):\n",
    "    valid_matrix_ids.add(matrix_id)\n",
    "\n",
    "# For each line...\n",
    "for line in Jglobals.parse_file(cisbp_tsv_file):\n",
    "\n",
    "    matrix_ids = re.findall(\"(M\\d{5}_2.00)\", line)\n",
    "\n",
    "    if matrix_ids:\n",
    "\n",
    "        line = line.strip(\"\\n\").split(\"\\t\")\n",
    "        gene_name = line[1]\n",
    "        species = line[2].replace(\"_\", \" \")\n",
    "\n",
    "        # Skip inferred TFs\n",
    "        if line[3] != \"D\":\n",
    "            continue\n",
    "\n",
    "        # For each TF...\n",
    "        for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "            if tf.gene_name.upper() == gene_name.upper() and species == tf.species:\n",
    "                tf.cisbp.update(valid_matrix_ids.intersection(set(matrix_ids)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ABF1 {'M00001_2.00', 'M00902_2.00'}\n",
      "ABF1 {'M01772_2.00', 'M01773_2.00'}\n",
      "ABF2 {'M00069_2.00'}\n",
      "ABF2 {'M01771_2.00'}\n",
      "ABF2\t841095.0\tArabidopsis thaliana\tQ9M7Q4\tAI5L5_ARATH\tReviewed\tMDGSMNLGNEPPGDGGGGGGLTRQGSIYSLTFDEFQSSVGKDFGSMNMDELLKNIWSAEETQAMASGVVPVLGGGQEGLQLQRQGSLTLPRTLSQKTVDQVWKDLSKVGSSGVGGSNLSQVAQAQSQSQSQRQQTLGEVTLEEFLVRAGVVREEAQVAARAQIAENNKGGYFGNDANTGFSVEFQQPSPRVVAAGVMGNLGAETANSLQVQGSSLPLNVNGARTTYQQSQQQQPIMPKQPGFGYGTQMGQLNSPGIRGGGLVGLGDQSLTNNVGFVQGASAAIPGALGVGAVSPVTPLSSEGIGKSNGDSSSLSPSPYMFNGGVRGRKSGTVEKVVERRQRRMIKNRESAARSRARKQAYTVELEAEVAKLKEENDELQRKQARIMEMQKNQETEMRNLLQGGPKKKLRRTESGPW\tbZIP\t\t2\tMA0941.1\t\t\t\t\t\tSRR2926839\t\tM01771_2.00\t\t\n",
      "...\n",
      "//\n",
      "Total genes: 1200\n",
      "Total feats: 1542\n"
     ]
    }
   ],
   "source": [
    "genes = 0\n",
    "feats = 0\n",
    "count = 0\n",
    "for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "    subtotal_feats = len(tf.cisbp)\n",
    "    if subtotal_feats > 0:\n",
    "        genes += 1\n",
    "        feats += subtotal_feats\n",
    "        if count <= 3:\n",
    "            print(tf.gene_name, tf.cisbp)\n",
    "        if count == 3:\n",
    "            print(tf)\n",
    "            print(\"...\")\n",
    "        count += 1\n",
    "print(\"//\")\n",
    "print(\"Total genes: %s\" % genes)\n",
    "print(\"Total feats: %s\" % feats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------#\n",
    "# UniPROBE    #\n",
    "#-------------#\n",
    "\n",
    "# For each line...\n",
    "for line in Jglobals.parse_tsv_file(uniprobe_file):\n",
    "   \n",
    "    if line[0] == \"Protein\":\n",
    "        continue\n",
    "\n",
    "    gene_name = line[0]\n",
    "    uniprobe_id = line[1]\n",
    "\n",
    "    # For each TF...\n",
    "    for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "        if tf.gene_name.upper() == gene_name.upper() and line[2] == tf.species:\n",
    "            tf.uniprobe.add(uniprobe_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ABF1 {'UP00452'}\n",
      "AFT1 {'UP00344'}\n",
      "ARO80 {'UP00329'}\n",
      "ARX {'UP00584'}\n",
      "ARX\t170302\tHomo sapiens\tQ96QS3\tARX_HUMAN\tReviewed\tMSNQYQEEGCSERPECKSKSPTLLSSYCIDSILGRRSPCKMRLLGAAQSLPAPLTSRADPEKAVQGSPKSSSAPFEAELHLPPKLRRLYGPGGGRLLQGAAAAAAAAAAAAAAAATATAGPRGEAPPPPPPTARPGERPDGAGAAAAAAAAAAAAWDTLKISQAPQVSISRSKSYRENGAPFVPPPPALDELGGPGGVTHPEERLGVAGGPGSAPAAGGGTGTEDDEEELLEDEEDEDEEEELLEDDEEELLEDDARALLKEPRRCPVAATGAVAAAAAAAVATEGGELSPKEELLLHPEDAEGKDGEDSVCLSAGSDSEEGLLKRKQRRYRTTFTSYQLEELERAFQKTHYPDVFTREELAMRLDLTEARVQVWFQNRRAKWRKREKAGAQTHPPGLPFPGPLSATHPLSPYLDASPFPPHHPALDSAWTAAAAAAAAAFPSLPPPPGSASLPPSGAPLGLSTFLGAAVFRHPAFISPAFGRLFSTMAPLTSASTAAALLRQPTPAVEGAVASGALADPATAAADRRASSIAALRLKAKEHAAQLTQLNILPGTSTGKEVC\tHomeodomain\t\t2\t\tARX_HUMAN.H11MO.0.D\t\t\t\t\t\tERR1002106;ERR1002107;ERR1002108;ERR1002109;ERR1002110;ERR1002111;ERR1002112;ERR1002113;ERR195529;ERR195530;ERR195531;ERR195532\tM00260_2.00;M00261_2.00;M00262_2.00;M00263_2.00;M00264_2.00;M00265_2.00\tUP00584\t\n",
      "...\n",
      "//\n",
      "Total genes: 541\n",
      "Total feats: 559\n"
     ]
    }
   ],
   "source": [
    "genes = 0\n",
    "feats = 0\n",
    "count = 0\n",
    "for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "    subtotal_feats = len(tf.uniprobe)\n",
    "    if subtotal_feats > 0:\n",
    "        genes += 1\n",
    "        feats += subtotal_feats\n",
    "        if count <= 3:\n",
    "            print(tf.gene_name, tf.uniprobe)\n",
    "        if count == 3:\n",
    "            print(tf)\n",
    "            print(\"...\")\n",
    "        else:\n",
    "            pass\n",
    "        count += 1\n",
    "print(\"//\")\n",
    "print(\"Total genes: %s\" % genes)\n",
    "print(\"Total feats: %s\" % feats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------#\n",
    "# SMiLE-seq   #\n",
    "#-------------#\n",
    "\n",
    "synonyms = {\n",
    "    \"CEBPb\": \"CEBPB\",\n",
    "    \"cFOS\": \"FOS\",\n",
    "    \"cFOSL2\": \"FOSL2\",\n",
    "    \"cJUN\": \"JUN\",\n",
    "    \"PPARa\": \"PPARA\",\n",
    "    \"PPARg\": \"PPARG\",\n",
    "    \"RXRa\": \"RXRA\",\n",
    "    \"RXRg\": \"RXRG\"            \n",
    "}\n",
    "\n",
    "# For each line...\n",
    "for line in Jglobals.parse_tsv_file(smile_seq):\n",
    "\n",
    "    if line[0] == \"Assay_Type\":\n",
    "        continue\n",
    "\n",
    "    sra_run = line[13]\n",
    "\n",
    "    for gene_name in line[16].split(\"_\")[0].split(\"-\"):\n",
    "        if gene_name in synonyms:\n",
    "            gene_name = synonyms[gene_name]\n",
    "\n",
    "        # For each TF...\n",
    "        for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "            if tf.gene_name.upper() == gene_name.upper() and line[12] == tf.species:\n",
    "                tf.smile_seq.add(sra_run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ARNTL {'SRR3405116', 'SRR3405117'}\n",
      "CEBPB {'SRR3405054'}\n",
      "CLOCK {'SRR3405116', 'SRR3405117'}\n",
      "CTCF {'SRR3405066', 'SRR3405055', 'SRR3405078'}\n",
      "CTCF\t10664\tHomo sapiens\tP49711\tCTCF_HUMAN\tReviewed\tMEGDAVEAIVEESETFIKGKERKTYQRRREGGQEEDACHLPQNQTDGGEVVQDVNSSVQMVMMEQLDPTLLQMKTEVMEGTVAPEAEAAVDDTQIITLQVVNMEEQPINIGELQLVQVPVPVTVPVATTSVEELQGAYENEVSKEGLAESEPMICHTLPLPEGFQVVKVGANGEVETLEQGELPPQEDPSWQKDPDYQPPAKKTKKTKKSKLRYTEEGKDVDVSVYDFEEEQQEGLLSEVNAEKVVGNMKPPKPTKIKKKGVKKTFQCELCSYTCPRRSNLDRHMKSHTDERPHKCHLCGRAFRTVTLLRNHLNTHTGTRPHKCPDCDMAFVTSGELVRHRRYKHTHEKPFKCSMCDYASVEVSKLKRHIRSHTGERPFQCSLCSYASRDTYKLKRHMRTHSGEKPYECYICHARFTQSGTMKMHILQKHTENVAKFHCPHCDTVIARKSDLGVHLRKQHSYIEQGKKCRYCDAVFHERYALIQHQKSHKNEKRFKCDQCDYACRQERHMIMHKRTHTGEKPYACSHCDKTFRQKQLLDMHFKRYHDPNFVPAAFVCSKCGKTFTRRNTMARHADNCAGPDGVEGENGGETKKSKRGRKRKMRSKKEDSSDSENAEPDLDDNEDEEEPAVEIEPEPEPQPVTPAPPPAKKRRGRPPGRTNQPKQNQPTAIIQVEDQNTGAIENIIVEVKKEPDAEPAEGEEEEAQPAATDAPNGDLTPEMILSMMDR\tC2H2 ZF\t\t3\tMA0139.1\tCTCF_HUMAN.H11MO.0.A\tERX008573;ERX008585;ERX008591;ERX008596;ERX008611;ERX1831867;ERX2868668;ERX2868669;ERX2868670;ERX2868671;ERX2868672;ERX2868673;ERX2868692;ERX2868693;ERX2868694;ERX2868695;ERX2868696;ERX2868697;ERX2868710;ERX2868711;ERX2868712;ERX2868713;ERX2868714;ERX2868715;ERX2868716;ERX2868717;ERX2868732;ERX2868733;ERX2868734;ERX2868735;ERX2868752;ERX2868761;ERX2868768;ERX2868777;ERX2868778;ERX2868782;ERX2868792;ERX2868793;ERX2868794;ERX2868800;ERX2868801;ERX2868811;ERX2868812;ERX2868813;ERX2868814;ERX2868821;ERX2868903;ERX2868904;ERX2868913;ERX2868914;ERX2869656;ERX2869660;ERX2869664;ERX2869668;SRX003794;SRX003796;SRX003798;SRX017278;SRX019498;SRX019505;SRX019512;SRX019519;SRX026719;SRX026720;SRX031244;SRX031245;SRX031246;SRX031247;SRX031248;SRX037935;SRX037936;SRX037937;SRX038504;SRX038505;SRX038524;SRX038525;SRX038543;SRX038544;SRX038562;SRX038563;SRX038582;SRX038583;SRX038602;SRX038603;SRX038622;SRX038623;SRX038644;SRX038645;SRX038664;SRX038665;SRX067387;SRX067396;SRX067423;SRX067446;SRX067467;SRX067470;SRX067475;SRX067495;SRX067503;SRX067513;SRX067516;SRX067534;SRX067535;SRX067536;SRX080142;SRX080143;SRX080328;SRX080329;SRX080330;SRX080331;SRX080333;SRX080336;SRX080337;SRX080338;SRX080339;SRX080340;SRX080341;SRX080342;SRX080343;SRX080344;SRX080346;SRX080347;SRX080349;SRX080350;SRX080351;SRX080352;SRX080353;SRX080355;SRX080356;SRX080357;SRX080358;SRX080359;SRX080362;SRX080367;SRX080368;SRX080369;SRX080370;SRX080371;SRX080373;SRX080374;SRX080375;SRX080377;SRX080378;SRX080380;SRX080386;SRX080388;SRX080389;SRX080390;SRX080391;SRX080392;SRX080393;SRX080395;SRX080396;SRX080398;SRX080399;SRX080400;SRX080402;SRX080403;SRX080404;SRX080406;SRX080408;SRX080410;SRX080411;SRX080412;SRX080413;SRX080414;SRX080415;SRX080416;SRX080420;SRX080422;SRX080425;SRX080427;SRX080431;SRX080432;SRX080442;SRX092414;SRX092416;SRX099863;SRX099864;SRX099870;SRX099871;SRX099875;SRX099876;SRX099882;SRX099883;SRX100378;SRX100393;SRX100446;SRX100464;SRX100498;SRX100501;SRX100531;SRX1026678;SRX1026679;SRX1026680;SRX1026681;SRX102973;SRX102977;SRX102978;SRX102979;SRX102980;SRX102984;SRX102986;SRX102988;SRX102993;SRX102996;SRX102998;SRX103002;SRX103004;SRX103007;SRX103008;SRX103010;SRX103011;SRX1030324;SRX1053364;SRX1053365;SRX1053373;SRX1053374;SRX1063404;SRX1063405;SRX1091813;SRX1091816;SRX1091817;SRX1091820;SRX1091821;SRX1091824;SRX1091825;SRX1091826;SRX111773;SRX111774;SRX111775;SRX111776;SRX111777;SRX111778;SRX1353402;SRX1353403;SRX1425809;SRX1430078;SRX150484;SRX150487;SRX150690;SRX151245;SRX1529431;SRX154498;SRX154499;SRX1594205;SRX1594206;SRX1594207;SRX1594208;SRX1594217;SRX1594218;SRX1738529;SRX1738531;SRX1744956;SRX1744957;SRX1744958;SRX1744959;SRX1744960;SRX1744961;SRX1801654;SRX1801655;SRX186611;SRX186638;SRX186660;SRX186670;SRX186704;SRX186777;SRX186778;SRX1878889;SRX1878890;SRX188946;SRX188947;SRX188950;SRX188951;SRX188952;SRX188955;SRX188958;SRX188959;SRX188960;SRX188962;SRX188963;SRX188964;SRX188968;SRX188970;SRX189083;SRX189084;SRX190190;SRX190230;SRX190276;SRX190359;SRX1950174;SRX1950175;SRX1950176;SRX1950177;SRX1967776;SRX1967777;SRX1967778;SRX1995217;SRX1995218;SRX1995219;SRX199854;SRX199856;SRX199857;SRX199858;SRX199859;SRX199861;SRX199862;SRX199863;SRX199864;SRX199865;SRX199867;SRX199868;SRX199871;SRX199872;SRX199878;SRX199879;SRX199880;SRX199881;SRX199885;SRX199886;SRX199889;SRX199890;SRX199891;SRX199892;SRX199893;SRX199894;SRX199895;SRX199896;SRX199897;SRX199899;SRX199903;SRX199904;SRX199905;SRX2012084;SRX212197;SRX212318;SRX212319;SRX2166890;SRX2166891;SRX2166892;SRX2166893;SRX2166894;SRX2166895;SRX219373;SRX2236973;SRX2442513;SRX2442515;SRX2442516;SRX2570608;SRX2570609;SRX2570610;SRX2570611;SRX265428;SRX2654818;SRX2654819;SRX2655459;SRX2655460;SRX2655461;SRX2655462;SRX2731772;SRX2731775;SRX2731778;SRX2896486;SRX3056625;SRX3056626;SRX3056628;SRX3056629;SRX3124393;SRX3124394;SRX3124407;SRX3124408;SRX3124419;SRX3124420;SRX3145148;SRX3172940;SRX3172941;SRX3288560;SRX3288561;SRX3288585;SRX3288586;SRX3367140;SRX3408800;SRX3433957;SRX3433959;SRX3433961;SRX3437414;SRX3437415;SRX3437416;SRX3437418;SRX3437419;SRX3437420;SRX3437421;SRX3437422;SRX3437423;SRX3437424;SRX3437425;SRX3437426;SRX3437427;SRX3437428;SRX3437429;SRX3437430;SRX3478024;SRX3478025;SRX3478026;SRX3478027;SRX3482872;SRX3482873;SRX3540868;SRX3540870;SRX356437;SRX356438;SRX356439;SRX356455;SRX356456;SRX356482;SRX356483;SRX356484;SRX356501;SRX356502;SRX356523;SRX356524;SRX356545;SRX356546;SRX356547;SRX356561;SRX356562;SRX356578;SRX356579;SRX356595;SRX356596;SRX356612;SRX356613;SRX356629;SRX356630;SRX356646;SRX356647;SRX356667;SRX356668;SRX356689;SRX356690;SRX356712;SRX356713;SRX356714;SRX356730;SRX356731;SRX356732;SRX356748;SRX356749;SRX356750;SRX356766;SRX356767;SRX356768;SRX356784;SRX356785;SRX356786;SRX356787;SRX3583180;SRX3583188;SRX3583196;SRX3583204;SRX3583212;SRX3583225;SRX359822;SRX359859;SRX359991;SRX360020;SRX360568;SRX360582;SRX3636814;SRX3636815;SRX3636816;SRX3636817;SRX3733779;SRX3733787;SRX3733796;SRX3733803;SRX3733807;SRX3800706;SRX3800707;SRX3800708;SRX3800709;SRX3800710;SRX3800711;SRX3800712;SRX381545;SRX3927292;SRX3927299;SRX3927305;SRX3927313;SRX395616;SRX395617;SRX4001754;SRX4001755;SRX4001756;SRX4001757;SRX4001758;SRX4001759;SRX4001760;SRX4001761;SRX4001833;SRX4001834;SRX4001879;SRX4001880;SRX4001881;SRX4001882;SRX4001883;SRX4001884;SRX4001885;SRX4001886;SRX4001887;SRX4001888;SRX4001889;SRX4001915;SRX4001916;SRX4001917;SRX4001918;SRX4001919;SRX4001920;SRX4001921;SRX4001922;SRX4001923;SRX4001924;SRX4001925;SRX4001946;SRX4001947;SRX4155650;SRX4155651;SRX4155662;SRX4155663;SRX4218025;SRX4221460;SRX4221464;SRX4284003;SRX4284004;SRX4284007;SRX4284008;SRX4284011;SRX4284012;SRX4288471;SRX4288472;SRX4288473;SRX4288474;SRX4288475;SRX4288476;SRX4374907;SRX4374908;SRX4377123;SRX4377124;SRX4377125;SRX4377126;SRX4377127;SRX4377128;SRX4377129;SRX4377130;SRX4377131;SRX4377132;SRX4377133;SRX4484978;SRX4521396;SRX4521397;SRX4521398;SRX4525850;SRX4547412;SRX4547413;SRX4547415;SRX4547416;SRX4655781;SRX4655782;SRX4655783;SRX4655784;SRX4655785;SRX4655786;SRX4655787;SRX4655788;SRX4714474;SRX4714479;SRX476480;SRX4792901;SRX4792902;SRX4792903;SRX4792904;SRX4792905;SRX4792906;SRX4890466;SRX4890467;SRX4958391;SRX4958392;SRX4958393;SRX4958394;SRX4958395;SRX4958396;SRX4958403;SRX4958404;SRX4958405;SRX4958406;SRX4958407;SRX4958408;SRX497255;SRX497256;SRX5001394;SRX5001395;SRX5001396;SRX5001397;SRX5126613;SRX5126616;SRX5126619;SRX5290903;SRX5290904;SRX5290909;SRX5290910;SRX5290911;SRX5290912;SRX5290913;SRX5290914;SRX5333219;SRX5333220;SRX5391686;SRX5391687;SRX5391688;SRX5391689;SRX5391690;SRX5391691;SRX539668;SRX5417099;SRX5522009;SRX5525074;SRX5806609;SRX5806610;SRX5848165;SRX5848166;SRX5848167;SRX5848171;SRX5848172;SRX5848173;SRX5848174;SRX5848175;SRX5848176;SRX5848177;SRX5848178;SRX5848179;SRX5848180;SRX5848181;SRX5848182;SRX5848183;SRX5883947;SRX5883948;SRX5883949;SRX5883950;SRX5883951;SRX5883952;SRX5927380;SRX5927381;SRX6578928;SRX6578929;SRX6578934;SRX6578935;SRX6578937;SRX6578938;SRX6578940;SRX6578941;SRX6578942;SRX6578943;SRX6578944;SRX6578946;SRX6578947;SRX6578948;SRX6680391;SRX6680397;SRX668216;SRX6759500;SRX6878423;SRX6878424;SRX6878425;SRX6878426;SRX6916365;SRX6916367;SRX6916369;SRX701956;SRX701962;SRX701963;SRX701964;SRX701965;SRX701966;SRX701967;SRX7058946;SRX7058947;SRX974395;SRX995496;SRX995500\t1203;1380;1381;1382;1383;1936;2254;2255;3049;3050;3051;3053;3054;3056;3057;3058;3059;3060;33297;33303;33554;33556;33561;33934;35380;35387;35392;35405;35406;35412;35416;35447;35498;35832;35840;35845;35846;35851;35854;35855;35924;35929;36573;36575;36578;36602;36610;36612;36616;36632;36636;36650;38185;38188;38192;38194;38196;38197;38203;38204;38774;39736;39737;39970;39971;39972;39973;39974;39975;40379;41283;42148;42149;42150;42151;42152;42153;42154;42155;42156;42157;42158;42159;42255;42256;42257;42271;42272;42296;42297;42298;42313;42314;42333;42334;42353;42354;42355;42369;42370;42384;42385;42399;42400;42414;42415;42429;42430;42444;42445;42463;42464;42483;42484;42504;42505;42506;42520;42521;42522;42536;42537;42538;42552;42553;42554;42568;42569;42570;42571;42650;42687;42848;42918;42932;43831;45144;45146;45178;45189;45218;45230;45279;45303;45323;45334;45347;45359;45395;45444;45455;45472;45494;45504;45526;45550;45551;45556;45558;45561;45562;45563;45564;45565;45566;45567;45568;45569;45570;45571;45572;45573;45574;45575;45576;45577;45578;45579;45581;45582;45584;45585;45588;45589;45590;45591;45592;45593;45594;45595;45596;45597;45598;45599;45600;45601;45602;45603;45604;45622;45623;45631;45645;45666;45667;45680;45703;45704;45705;45706;45707;45708;45709;45710;45711;45712;45713;45716;45717;45718;45719;45720;45721;45733;45734;45760;45794;45795;45808;45832;45833;45834;45835;45836;45837;45838;45839;45840;45841;45842;45843;45844;45845;45846;45847;45850;45851;45855;45858;45859;45865;45899;45908;45912;45913;45974;45975;45976;45977;45983;45984;45987;45988;45992;45994;45995;46009;46013;46014;46015;46016;46017;46018;46023;46028;46029;46030;46031;46035;46042;46043;46044;46047;46048;46049;46050;46053;46063;46091;46184;46218;46250;46271;46318;46367;46373;47281;47474;47475;47476;47477;47776;47896;47897;48481;48482;48757;49479;5039;5040;5041;5042;5043;5044;5045;5046;5047;5048;5049;5050;5051;50987;50988;50989;50990;50991;51438;51439;51441;51442;51443;51444;53885;54151;55727;55728;55770;55773;55774;55777;55778;55781;55782;55783;57283;57284;5828;5830;58360;58361;58362;58363;58480;58481;58489;58490;62569;62586;62707;62852;62917;62979;63059;63245;63284;63328;63479;63480;63691;63753;63868;63899;64040;64041;64044;64047;64048;64159;64259;64413;64428;64457;64488;64528;64614;65090;65092;65093;65094;65095;65096;65097;6572;6573;6574;6575;6576;67699;69619;69621;69624;69627;70062;70775;71103;71104;71105;71106;71669;71670;71677;71678;71679;71680;72362;72779;72780;72781;72977;72978;73000;73001;73002;74902;74903;74904;74905;75553;75554;75555;75556;75557;75558;75572;75573;75574;75575;75847;75848;76247;76248;76249;76250;76251;76252;8086;8087;8088;8089;8090;8103;8104;81228;81229;81659;81660;81661;81662;81663;81664;82007;82974;82975;82982;832;83249;83251;83252;83253;83591;83592;83595;83736;83737;83738;83739;83740;83741;83742;84200;85008;85009;85284;85285;85286;85287;85639;85640;85641;85680;85767;85815;85819;85826;85870;86416;86441;86828;87100;87102;87487;87488;87644;87645;87646;87647;87650;87651;87830;88132;88275;88815;88917;9316\tEXP036890;EXP037459;EXP037532;EXP037799;EXP037800;EXP037801;EXP037952;EXP038267;EXP038268;EXP038623;EXP038624;EXP039310;EXP039321;EXP039328;EXP039334;EXP039370;EXP039379;EXP039386;EXP039388;EXP039398;EXP039402;EXP039406;EXP039407;EXP039413;EXP039434;EXP039438;EXP039453;EXP039454;EXP039457;EXP039463;EXP039468;EXP039469;EXP039472;EXP039474;EXP039475;EXP039487;EXP039498;EXP039523;EXP039534;EXP039536;EXP039537;EXP039544;EXP039547;EXP039548;EXP039550;EXP039569;EXP039572;EXP039574;EXP039576;EXP039579;EXP039581;EXP039586;EXP039609;EXP039610;EXP039612;EXP039618;EXP039620;EXP039638;EXP039639;EXP039645;EXP039663;EXP039667;EXP039677;EXP039678;EXP039695;EXP039696;EXP039699;EXP039700;EXP039704;EXP039705;EXP039707;EXP039723;EXP039740;EXP039745;EXP039759;EXP039762;EXP039772;EXP039781;EXP039782;EXP039802;EXP039805;EXP039806;EXP039832;EXP039840;EXP039845;EXP039854;EXP039861;EXP039864;EXP039866;EXP039880;EXP039898;EXP039929;EXP039943;EXP039949;EXP039952;EXP039954;EXP039968;EXP039971;EXP039980;EXP039983;EXP039994;EXP040000;EXP040012;EXP040026;EXP040028;EXP040031;EXP040050;EXP040054;EXP040061;EXP040063;EXP040068;EXP040071;EXP040072;EXP040077;EXP040082;EXP040087;EXP040089;EXP040101;EXP040107;EXP040118;EXP040128;EXP040134;EXP040145;EXP040169;EXP040192;EXP040193;EXP040228;EXP040241;EXP040244;EXP040246;EXP040253;EXP040254;EXP040256;EXP040257;EXP040275;EXP040279;EXP040282;EXP040290;EXP040291;EXP040298;EXP040321;EXP040383;EXP040384;EXP040385;EXP040425;EXP040426;EXP040768\tENCSR000AHD.CTCF.MCF-7;ENCSR000AKB.CTCF.GM12878;ENCSR000AKO.CTCF.K-562;ENCSR000ALA.CTCF.endothelial_umbilical-vein;ENCSR000ALJ.CTCF.keratinocyte;ENCSR000ALV.CTCF.epithelial_mammary;ENCSR000AMA.CTCF.Hep-G2;ENCSR000AMF.CTCF.WA01;ENCSR000ANE.CTCF.myoblast_skeletal_muscle;ENCSR000ANO.CTCF.fibroblast_LUNG;ENCSR000ANS.CTCF.myotube;ENCSR000AOA.CTCF.HeLa-S3;ENCSR000AOO.CTCF.astrocyte;ENCSR000APF.CTCF.osteoblast;ENCSR000APM.CTCF.fibroblast_DERMAL;ENCSR000AQU.CTCF.DND41;ENCSR000ATN.CTCF.CD14;ENCSR000AUE.CTCF.A-549;ENCSR000AUF.CTCF.A-549;ENCSR000AUV.CTCF.B-cell;ENCSR000BHV.CTCF.A-549;ENCSR000BHW.CTCF.A-549;ENCSR000BIE.CTCF.Hep-G2;ENCSR000BLX.CTCF.SK-N-SH;ENCSR000BNH.CTCF.WA01;ENCSR000BNO.CTCF.T-47D;ENCSR000BPJ.CTCF.K-562;ENCSR000BQE.CTCF.Ishikawa;ENCSR000BSE.CTCF.HCT-116;ENCSR000DKN.CTCF.H54;ENCSR000DKP.CTCF.GM10248;ENCSR000DKR.CTCF.GM10266;ENCSR000DKV.CTCF.GM12878;ENCSR000DKX.CTCF.GM12891;ENCSR000DKY.CTCF.GM12892;ENCSR000DKZ.CTCF.GM13976;ENCSR000DLB.CTCF.GM13977;ENCSR000DLD.CTCF.GM19238;ENCSR000DLE.CTCF.GM19239;ENCSR000DLF.CTCF.GM19240;ENCSR000DLG.CTCF.GM20000;ENCSR000DLK.CTCF.WA01;ENCSR000DLO.CTCF.HeLa-S3;ENCSR000DLS.CTCF.Hep-G2;ENCSR000DLW.CTCF.endothelial_umbilical-vein;ENCSR000DMA.CTCF.K-562;ENCSR000DMC.CTCF.kidney;ENCSR000DME.CTCF.LNCaP;ENCSR000DMF.CTCF.LNCaP;ENCSR000DMH.CTCF.lung;ENCSR000DML.CTCF.MCF-7;ENCSR000DMO.CTCF.MCF-7;ENCSR000DMR.CTCF.MCF-7;ENCSR000DMS.CTCF.MCF-7;ENCSR000DMV.CTCF.MCF-7;ENCSR000DMY.CTCF.medulloblastoma;ENCSR000DNA.CTCF.A-549;ENCSR000DNC.CTCF.keratinocyte;ENCSR000DND.CTCF.pancreas;ENCSR000DNI.CTCF.spleen;ENCSR000DPF.CTCF.A-549;ENCSR000DPG.CTCF.fibroblast_SKIN_LEG;ENCSR000DPM.CTCF.fibroblast_LUNG;ENCSR000DPP.CTCF.fibroblast_PEDAL_DIGIT_SKIN;ENCSR000DPS.CTCF.fibroblast_GINGIVA;ENCSR000DPV.CTCF.fibroblast_SKIN_ABDOMEN;ENCSR000DPY.CTCF.fibroblast_AORTIC_ADVENTITIA;ENCSR000DQD.CTCF.BE2C;ENCSR000DQI.CTCF.BJ;ENCSR000DQN.CTCF.CACO2;ENCSR000DQW.CTCF.GM06990;ENCSR000DQY.CTCF.GM12801;ENCSR000DRB.CTCF.GM12864;ENCSR000DRE.CTCF.GM12865;ENCSR000DRF.CTCF.GM12866;ENCSR000DRH.CTCF.GM12867;ENCSR000DRI.CTCF.GM12868;ENCSR000DRJ.CTCF.GM12869;ENCSR000DRK.CTCF.GM12870;ENCSR000DRL.CTCF.GM12871;ENCSR000DRN.CTCF.GM12872;ENCSR000DRP.CTCF.GM12873;ENCSR000DRR.CTCF.GM12874;ENCSR000DRU.CTCF.GM12875;ENCSR000DRZ.CTCF.GM12878;ENCSR000DSU.CTCF.astrocyte_spinal_cord;ENCSR000DSZ.CTCF.astrocyte_cerebellum;ENCSR000DTA.CTCF.endothelial_brain-microvascular;ENCSR000DTF.CTCF.fibroblast_CARDIAC;ENCSR000DTI.CTCF.cardiac_muscle;ENCSR000DTL.CTCF.choroid-plexus_epithelial;ENCSR000DTO.CTCF.HCT-116;ENCSR000DTR.CTCF.epithelial_esophagus;ENCSR000DTW.CTCF.HEK293;ENCSR000DUB.CTCF.HeLa-S3;ENCSR000DUG.CTCF.Hep-G2;ENCSR000DUH.CTCF.fibroblast_FORESKIN;ENCSR000DUM.CTCF.HFF;ENCSR000DUP.CTCF.HL-60;ENCSR000DUS.CTCF.epithelial_mammary;ENCSR000DUU.CTCF.fibroblast_MAMMARY;ENCSR000DUX.CTCF.fibroblast_PULMONARY_ARTERY;ENCSR000DVA.CTCF.fibroblast_LUNG;ENCSR000DVH.CTCF.epithelial_kidney;ENCSR000DVI.CTCF.epithelial_retinal_pigment;ENCSR000DVP.CTCF.endothelial_umbilical-vein;ENCSR000DVQ.CTCF.fibroblast_VILLOUS_MESENCHYME;ENCSR000DWE.CTCF.K-562;ENCSR000DWH.CTCF.MCF-7;ENCSR000DWN.CTCF.NB4;ENCSR000DWQ.CTCF.fibroblast_FORESKIN;ENCSR000DWX.CTCF.keratinocyte;ENCSR000DWY.CTCF.fibroblast_LUNG;ENCSR000DXD.CTCF.epithelial_proximal_tubule;ENCSR000DXI.CTCF.bonchial_epithelial;ENCSR000DXQ.CTCF.SK-N-SH;ENCSR000DXW.CTCF.WERI-Rb-1;ENCSR000DYB.CTCF.WI-38;ENCSR000DYD.CTCF.A-549;ENCSR000DZN.CTCF.GM12878;ENCSR000EFI.CTCF.IMR-90;ENCSR000EGM.CTCF.K-562;ENCSR000EIC.CTCF.SK-N-SH;ENCSR003SZZ.CTCF.esophagus_squamous-epithelium;ENCSR014GSQ.CTCF.adrenal-gland;ENCSR027FSZ.CTCF.lung_left_upper-lobe;ENCSR027HML.CTCF.OCI-Ly7;ENCSR028YEV.CTCF.spleen;ENCSR038FOS.CTCF.skin_suprapubic;ENCSR066BZZ.CTCF.neuron_bipolar_doxy_4d;ENCSR066GBX.CTCF.cardiac_right-atrium-auricular-region;ENCSR071XWO.CTCF.gastrocnemius-medialis;ENCSR072EUE.CTCF.OCI-Ly1;ENCSR074SFL.CTCF.esophagus_muscularis-mucosa;ENCSR084RDK.CTCF.DOHH2;ENCSR089DTY.CTCF.omental-fat-pad;ENCSR102CSD.CTCF.colon_transverse;ENCSR113REG.CTCF.KARPAS422;ENCSR125DKL.CTCF.SU-DHL-6;ENCSR125NBL.CTCF.neural_progenitor;ENCSR145HBC.CTCF.thyroid-gland;ENCSR146BGM.CTCF.gastroesophageal-sphincter;ENCSR173AIR.CTCF.stomach;ENCSR175FLL.CTCF.coronary-artery;ENCSR186NVR.CTCF.gastroesophageal-sphincter;ENCSR196HOM.CTCF.prostate_epithelial;ENCSR203QEB.CTCF.PANC-1;ENCSR206ETG.CTCF.gastroesophageal-sphincter;ENCSR222SQE.CTCF.colon_sigmoid;ENCSR225OKX.CTCF.omental-fat-pad;ENCSR234HEM.CTCF.spleen;ENCSR236YGF.CTCF.colon_transverse;ENCSR240PRQ.CTCF.HCT-116;ENCSR252QYR.CTCF.hepatocyte;ENCSR252XWG.CTCF.skin_lower-leg;ENCSR254YRM.CTCF.liver;ENCSR261VAS.CTCF.smooth-muscle-cell;ENCSR265ARE.CTCF.VCaP;ENCSR265PFQ.CTCF.pancreas_body;ENCSR303GFI.CTCF.RWPE-1;ENCSR304XUZ.CTCF.breast_epithelium;ENCSR307PFP.CTCF.pancreas_body;ENCSR315NAC.CTCF.LNCaP;ENCSR353DFU.CTCF.esophagus_muscularis-mucosa;ENCSR359LOD.CTCF.PC-3;ENCSR361KVZ.CTCF.stomach;ENCSR375VXU.CTCF.Peyers-patch;ENCSR391ZKN.CTCF.Peyers-patch;ENCSR392SFJ.CTCF.uterus;ENCSR402IDP.CTCF.MM1-S;ENCSR408XTO.CTCF.pancreas_body;ENCSR408ZEE.CTCF.adrenal-gland;ENCSR419ANE.CTCF.Peyers-patch;ENCSR428BKN.CTCF.gastrocnemius-medialis;ENCSR430TEE.CTCF.skin_lower-leg;ENCSR434XLP.CTCF.nerve_tibial;ENCSR449SEF.CTCF.colon_transverse;ENCSR450BLH.CTCF.adrenal-gland;ENCSR450FRI.CTCF.esophagus_squamous-epithelium;ENCSR460LGH.CTCF.LNCaP-C4-2B;ENCSR464DKE.CTCF.Loucy;ENCSR469POZ.CTCF.nerve_tibial;ENCSR482PMN.CTCF.spleen;ENCSR484DDO.CTCF.pancreas_body;ENCSR492ZIW.CTCF.thyroid-gland;ENCSR493APD.CTCF.ovary;ENCSR494TNM.CTCF.testis;ENCSR515LRI.CTCF.skin_suprapubic;ENCSR527TPP.CTCF.uterus;ENCSR541AMF.CTCF.SK-N-SH;ENCSR544APK.CTCF.heart_left-ventricle;ENCSR548DDS.CTCF.ovary;ENCSR549WAU.CTCF.stomach;ENCSR558HTE.CTCF.colon_transverse;ENCSR559KAB.CTCF.esophagus_muscularis-mucosa;ENCSR560BUE.CTCF.MCF-7;ENCSR572DUJ.CTCF.pancreas_body;ENCSR582MTM.CTCF.skin_lower-leg;ENCSR606TNN.CTCF.vagina;ENCSR608WPS.CTCF.colon_transverse;ENCSR611JJS.CTCF.A-673;ENCSR617IFZ.CTCF.HEK293;ENCSR618QYE.CTCF.stomach;ENCSR619IUE.CTCF.neuron_bipolar_doxy_4d;ENCSR634OAQ.CTCF.NCI-H929;ENCSR655ECZ.CTCF.vagina;ENCSR661NXJ.CTCF.breast_epithelium;ENCSR661XNQ.CTCF.gastroesophageal-sphincter;ENCSR668BTN.CTCF.aorta_thoracic;ENCSR673WZL.CTCF.LNCaP_hydroxy_10nM_4h;ENCSR692ILH.CTCF.spleen;ENCSR697YIN.CTCF.breast_epithelium;ENCSR699BEK.CTCF.artery_tibial;ENCSR713SXF.CTCF.cardiac-muscle;ENCSR718SDR.CTCF.heart_left-ventricle;ENCSR720USO.CTCF.prostate_gland;ENCSR721AHD.CTCF.colon_sigmoid;ENCSR744YJR.CTCF.thyroid-gland;ENCSR756KRS.CTCF.skin_suprapubic;ENCSR756URL.CTCF.esophagus_squamous-epithelium;ENCSR756ZKG.CTCF.OCI-Ly3;ENCSR762PCG.CTCF.subcutaneous-adipose-tissue;ENCSR769WKR.CTCF.colon_transverse;ENCSR770IWO.CTCF.adrenal-gland;ENCSR773JBP.CTCF.esophagus_squamous-epithelium;ENCSR785YRL.CTCF.neutrophil;ENCSR791AYW.CTCF.heart_left-ventricle;ENCSR798NVH.CTCF.uterus;ENCSR799TJD.CTCF.lung_left_upper-lobe;ENCSR799WDT.CTCF.Peyers-patch;ENCSR822CEA.CTCF.neural;ENCSR822PJT.CTCF.nerve_tibial;ENCSR829HTO.CTCF.prostate_gland;ENCSR833FWC.CTCF.colon_transverse;ENCSR838RUX.CTCF.esophagus_squamous-epithelium;ENCSR847XGE.CTCF.22Rv1_hydroxy_10nM_4h;ENCSR856JJB.CTCF.RWPE-2;ENCSR857PBV.CTCF.22Rv1;ENCSR899JSO.CTCF.adrenal-gland;ENCSR907BES.CTCF.colon_transverse;ENCSR911GFJ.CTCF.liver_right-lobe-of;ENCSR925GDS.CTCF.colon_sigmoid;ENCSR932ZMX.CTCF.GM23338;ENCSR955BIB.CTCF.thyroid-gland;ENCSR960MDF.CTCF.aorta_ascending;ENCSR964BKO.CTCF.lung_left_upper-lobe;ENCSR970UZD.CTCF.lung_left_upper-lobe;ENCSR972LYL.CTCF.lung_left_upper-lobe;ENCSR981CID.CTCF.testis;ENCSR987GXT.CTCF.GM23338;ENCSR998NQG.CTCF.gastrocnemius-medialis;ERP000209.CTCF.Hep-G2;ERP000209.CTCF.MCF-7_E2_SHCTCF;ERP000209.CTCF.MCF-7_E2_SHCTR;ERP000380.CTCF.MCF-7;ERP000380.CTCF.MCF-7_E2;ERP000380.CTCF.MCF-7_TAM;ERP000395.CTCF.hepatocyte;ERP000783.CTCF.MCF-10A;ERP000783.CTCF.MCF-7_TAM;ERP000783.CTCF.ZR751;ERP002246.CTCF.81-3;ERP002246.CTCF.GM12878;ERP004003.CTCF.islet;ERP008568.CTCF.CHRF28811;ERP008568.CTCF.HL-60;ERP008568.CTCF.U-937;GSE101295.CTCF.NB69;GSE101295.CTCF.SH-SY5Y;GSE101295.CTCF.SK-N-AS;GSE102184.CTCF.erythroblast_adult_erythroblasts;GSE102184.CTCF.fetal_erythroblasts;GSE102237.CTCF.226LDM;GSE102884.CTCF.HeLa-Kyoto;GSE102884.CTCF.HeLa-Kyoto_PDS5-depleted;GSE102884.CTCF.HeLa-Kyoto_WAPL-depleted;GSE102884.CTCF.HeLa-Kyoto_WAPL_PDS-depleted;GSE103477.CTCF.MDM;GSE103477.CTCF.MDM_H5N1;GSE103477.CTCF.MDM_IFNb;GSE103477.CTCF.MDM_dNS1;GSE103477.CTCF.THP-1_PMA_Dex-0h;GSE103477.CTCF.THP-1_PMA_Dex-6h;GSE103477.CTCF.THP-1_Pam3csk-000m-Flavo-000m;GSE103477.CTCF.THP-1_Pam3csk-000m-Flavo-240m;GSE103477.CTCF.THP-1_Pam3csk-020m-Flavo-000m;GSE103477.CTCF.THP-1_Pam3csk-025m-Flavo-000m;GSE103477.CTCF.THP-1_Pam3csk-030m-Flavo-000m;GSE103477.CTCF.THP-1_Pam3csk-045m-Flavo-000m;GSE103477.CTCF.THP-1_Pam3csk-060m-Flavo-000m;GSE103477.CTCF.THP-1_Pam3csk-120m-Flavo-000m;GSE103477.CTCF.THP-1_Pam3csk-150m-Flavo-030m;GSE103477.CTCF.THP-1_Pam3csk-180m-Flavo-060m;GSE103477.CTCF.THP-1_Pam3csk-360m-Flavo-240m;GSE103477.CTCF.THP-1_eGFP-IFNb-r1;GSE103477.CTCF.THP-1_siCtrl-NS1-Pam3csk-0h;GSE103477.CTCF.THP-1_siCtrl-NS1-Pam3csk-4h;GSE103477.CTCF.THP-1_siCtrl-NS1-Pam3csk-7h-Flavo-3h;GSE103477.CTCF.THP-1_siCtrl-eGFP-Pam3csk-0h;GSE103477.CTCF.THP-1_siCtrl-eGFP-Pam3csk-4h;GSE103477.CTCF.THP-1_siNIPBL-NS1-Pam3csk-4h;GSE103477.CTCF.THP-1_siNIPBL-NS1-Pam3csk-7h-Flavo-3h;GSE103477.CTCF.THP-1_siNIPBL-eGFP-Pam3csk-4h;GSE103477.CTCF.THP-1_siWAPL-NS1-Pam3csk-4h;GSE103477.CTCF.THP-1_siWAPL-NS1-Pam3csk-7h-Flavo-3h;GSE103477.CTCF.THP-1_siWAPL-eGFP-Pam3csk-4h;GSE103651.CTCF.HEK293T;GSE104676.CTCF.HUDEP-2_30min;GSE104888.CTCF.HCT-116_RAD21-mAC;GSE104888.CTCF.HCT-116_RAD21-mAC_500uM_auxin;GSE105028.CTCF.WA09;GSE105028.CTCF.WA09_heat-shock;GSE107147.CTCF.HSPC-CD34;GSE107607.CTCF.HeLa_dC9Sun-D3AMut_MIR152;GSE107607.CTCF.HeLa_dC9Sun-D3A_CDCC85C;GSE107607.CTCF.HeLa_dC9Sun-D3A_MIR152;GSE107607.CTCF.HeLa_dC9Sun-D3A_SHB;GSE107749.CTCF.MCF-7;GSE107749.CTCF.MCF-7_estrogen_45min;GSE108173.CTCF.HeLa-S3_synchro;GSE108173.CTCF.HeLa-S3_unsynchro;GSE109706.CTCF.LUHMES;GSE110681.CTCF.K-562;GSE111000.CTCF.Hep-G2;GSE111000.CTCF.Hep-G2_RELACS;GSE111000.CTCF.Hep-G2_RELACS_10000cells;GSE111170.CTCF.GM04025_B-lymphocytes;GSE111170.CTCF.GM09236__B-lymphocytes;GSE111170.CTCF.GM09237_B-lymphocytes;GSE118305.CTCF.macrophage;GSE118305.CTCF.macrophage_ZIKVneg;GSE118305.CTCF.macrophage_ZIKVnpos;GSE20650.CTCF.hESC;GSE21366.CTCF.ASC;GSE23784.CTCF.islet;GSE31485.CTCF.BJAB;GSE31485.CTCF.BL41;GSE38411.CTCF.BCBL-1;GSE41048.CTCF.WI-38VA13;GSE43070.CTCF.IMR-90;GSE43098.CTCF.monocyte_BLOOD;GSE43098.CTCF.monocyte_MACROPHAGE;GSE43743.CTCF.MM1-S;GSE46055.CTCF.SUM159;GSE46832.CTCF.lymphocyte_blood;GSE51234.CTCF.GP5D;GSE51234.CTCF.GP5D_SIRAD21;GSE52457.CTCF.WA01_EMB;GSE52457.CTCF.WA01_MESEN;GSE52457.CTCF.WA01_MESEN_STEM;GSE52457.CTCF.WA01_NEUR;GSE52457.CTCF.WA01_TROP;GSE67783.CTCF.HSPC;GSE67783.CTCF.erythroid;GSE68976.CTCF.HEK293;GSE68976.CTCF.Jurkat;GSE69646.CTCF.hESC_NAIVE;GSE69646.CTCF.hESC_PRIMED;GSE69962.CTCF.THP-1;GSE69962.CTCF.THP-1_2D3;GSE70482.CTCF.K-562;GSE70764.CTCF.K-562;GSE70764.CTCF.MCF-7;GSE70764.CTCF.OVCAR-8;GSE70764.CTCF.delta-47;GSE74850.CTCF.tlymphocyte_naive;GSE76494.CTCF.HEK293;GSE76815.CTCF.SK-N-SH;GSE76922.CTCF.GM12878;GSE76922.CTCF.GM17942;GSE76922.CTCF.ID00014;GSE76922.CTCF.ID00015;GSE76922.CTCF.ID00016;GSE78113.CTCF.786-O_NORMOXIA;GSE78113.CTCF.MCF-7_HYPOXIA;GSE78113.CTCF.MCF-7_NORMOXIA;GSE78957.CTCF.IMR-5;GSE81945.CTCF.COLO-829;GSE83726.CTCF.RH4;GSE84432.CTCF.VCaP_ETOH;GSE84432.CTCF.VCaP_R1881;GSE84657.CTCF.primary-epidermal-keratinocyte_diff_d0;GSE84657.CTCF.primary-epidermal-keratinocyte_diff_d3;GSE84657.CTCF.primary-epidermal-keratinocyte_diff_d6;GSE86981.CTCF.retina_AB1-FW14;GSE86981.CTCF.retina_AB1-FW16;GSE86981.CTCF.retina_AB1-FW18;GSE86981.CTCF.retina_AB1-FW23;GSE86981.CTCF.retina_AB1-RB;GSE87831.CTCF.U2OS;GSE88734.CTCF.MIA-PaCa-2;GSE90691.CTCF.K-562_HOXA11_MQ1Q147L;GSE90691.CTCF.K-562_HOXA11_dMQ1;GSE90691.CTCF.K-562_HOXA13_MQ1Q147L;GSE90691.CTCF.K-562_HOXA13_dMQ1;GSE90691.CTCF.K-562_RUNX1_MQ1Q147L;GSE90691.CTCF.K-562_RUNX1_dMQ1;GSE92879.CTCF.HCT-116;GSE92879.CTCF.K-562;GSE92879.CTCF.K-562_Dox;GSE94488.CTCF.SK-MEL-147;GSE94992.CTCF.HAP1;GSE94992.CTCF.HAP1_SCC4KO;GSE94992.CTCF.HAP1_WaplKO-33;GSE94992.CTCF.HAP1_WaplKO-33_SCC4KO;GSE96800.CTCF.THP-1_macrophage_PMA;GSE96800.CTCF.THP-1_monocytes;GSE97394.CTCF.HUES-64;GSE97394.CTCF.HUES-64_DNMT-KO;GSE98551.CTCF.MCF-10A;GSE98551.CTCF.MCF-10AT1;GSE98551.CTCF.MCF-10CA1a\t\tERR173154;ERR173155;ERR173156;ERR173157\t\t\tSRR3405055;SRR3405066;SRR3405078\n",
      "...\n",
      "//\n",
      "Total genes: 70\n",
      "Total feats: 151\n"
     ]
    }
   ],
   "source": [
    "genes = 0\n",
    "feats = 0\n",
    "count = 0\n",
    "for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "    subtotal_feats = len(tf.smile_seq)\n",
    "    if subtotal_feats > 0:\n",
    "        genes += 1\n",
    "        feats += subtotal_feats\n",
    "        if count <= 3:\n",
    "            print(tf.gene_name, tf.smile_seq)\n",
    "        if count == 3:\n",
    "            print(tf)\n",
    "            print(\"...\")\n",
    "        else:\n",
    "            pass\n",
    "        count += 1\n",
    "print(\"//\")\n",
    "print(\"Total genes: %s\" % genes)\n",
    "print(\"Total feats: %s\" % feats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ofornes/Anaconda3/lib/python3.8/site-packages/Bio/SubsMat/__init__.py:126: BiopythonDeprecationWarning: Bio.SubsMat has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.substitution_matrices as a replacement, and contact the Biopython developers if you still need the Bio.SubsMat module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "#-------------#\n",
    "# Pfam        #\n",
    "#-------------#\n",
    "\n",
    "# The following code is adapted from:\n",
    "# https://github.com/wassermanlab/JASPAR-profile-inference/blob/master/files/get_files.py\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import json\n",
    "from infer_profile import hmmAlign, hmmScan, _makeSeqFile\n",
    "\n",
    "# Initialize\n",
    "seq_file = \".seq.fasta\"\n",
    "hmm_database = os.path.join(jaspar_dir, \"files\", \"pfam-DBDs\", \"all_DBDs.hmm\")\n",
    "\n",
    "# Change dir\n",
    "os.chdir(sys.path[0])\n",
    "os.chdir(os.path.abspath(\"./Data/Pfam/\"))\n",
    "\n",
    "# Skip if JSON file already exists\n",
    "json_file = \"TFs.json\"\n",
    "if not os.path.exists(json_file):\n",
    "\n",
    "    # Initialize\n",
    "    pfams = {}\n",
    "\n",
    "    # For each TF...\n",
    "    for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "\n",
    "        # Initialize\n",
    "        unientry = tf.unientry\n",
    "        pfams.setdefault(unientry, [])\n",
    "\n",
    "        # Make seq file\n",
    "        seq = Seq(tf.sequence)\n",
    "        seq_record = SeqRecord(seq, id=unientry, name=unientry, description=unientry)\n",
    "        _makeSeqFile(seq_record, seq_file)\n",
    "\n",
    "        # For each DBD...\n",
    "        for pfam_ac, start, end, evalue in hmmScan(seq_file, hmm_database, non_overlapping_domains=True):\n",
    "\n",
    "            # Initialize\n",
    "            hmm_file = os.path.join(jaspar_dir, \"files\", \"pfam-DBDs\", \"%s.hmm\" % pfam_ac)\n",
    "\n",
    "            # Make seq file\n",
    "            sub_seq = seq[start:end]\n",
    "            seq_record = SeqRecord(sub_seq, id=unientry, name=unientry, description=unientry)\n",
    "            _makeSeqFile(seq_record, seq_file)\n",
    "\n",
    "            # Add DBDs\n",
    "            alignment = hmmAlign(seq_file, hmm_file)\n",
    "            pfams[unientry].append((pfam_ac, alignment, start+1, end, evalue))\n",
    "\n",
    "    # Write\n",
    "    Jglobals.write(json_file, json.dumps(pfams, sort_keys=True, indent=4, separators=(\",\", \": \")))\n",
    "\n",
    "    # Remove seq file\n",
    "    if os.path.exists(seq_file):\n",
    "        os.remove(seq_file)\n",
    "\n",
    "# Change dir back\n",
    "os.chdir(sys.path[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------#\n",
    "# Cluster     #\n",
    "# (MMseqs2)   #\n",
    "#-------------#\n",
    "\n",
    "# Get species\n",
    "species = {}\n",
    "for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "    species.setdefault(tf.species, tf.species.replace(\" \", \"_\"))\n",
    "\n",
    "# Get species\n",
    "species = {}\n",
    "for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "    species.setdefault(tf.species, tf.species.replace(\" \", \"_\"))\n",
    "\n",
    "# For each species...\n",
    "for s in sorted(species):\n",
    "\n",
    "    # Initialize\n",
    "    cluster_num = 0\n",
    "    clusters = {}\n",
    "    cluster_ids = set()\n",
    "    clusters_dir = \"./Data/Clusters/%s\" % species[s]\n",
    "    family_num = 0\n",
    "    families = {}\n",
    "    mmseqs_path = \"/Users/ofornes/.anaconda3/bin/mmseqs\"\n",
    "\n",
    "\n",
    "    if not os.path.isdir(clusters_dir):\n",
    "        os.mkdir(clusters_dir)\n",
    "\n",
    "    # Get families\n",
    "    for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "        if tf.species != s:\n",
    "            continue\n",
    "        if tf.family not in families:\n",
    "            family_num += 1\n",
    "            families.setdefault(tf.family, family_num)\n",
    "            prefix = os.path.join(clusters_dir, str(families[tf.family]))\n",
    "            fasta_file = \"%s.fasta\" % prefix\n",
    "            if os.path.exists(fasta_file):\n",
    "                os.remove(fasta_file)\n",
    "\n",
    "    # Create family-specific FASTA files\n",
    "    for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "        if tf.species != s:\n",
    "            continue\n",
    "        prefix = os.path.join(clusters_dir, str(families[tf.family]))\n",
    "        fasta_file = \"%s.fasta\" % prefix\n",
    "        Jglobals.write(fasta_file, \">%s\\n%s\" % (tf.unientry, tf.sequence))\n",
    "\n",
    "    # Skip if JSON file already exists\n",
    "    json_file = \"%s/%s.json\" % (clusters_dir, species[s])\n",
    "    if not os.path.exists(json_file):\n",
    "\n",
    "        for family in sorted(families.values()):\n",
    "\n",
    "            # Initialize\n",
    "            prefix = os.path.join(clusters_dir, str(family))\n",
    "            cluster_all_seqs = \"%s_all_seqs.fasta\" % prefix\n",
    "            cluster_clusters = \"%s_cluster.tsv\" % prefix\n",
    "            cluster_rep_seqs = \"%s_rep_seq.fasta\" % prefix\n",
    "\n",
    "            # Cluster TFs\n",
    "            if not os.path.exists(cluster_rep_seqs):\n",
    "                fasta_file = \"%s.fasta\" % prefix\n",
    "                # From Burkhard Rost's \"Twilight zone of protein sequence alignments\":\n",
    "                # Sequence alignments unambiguously distinguish between protein pairs of similar and\n",
    "                # non-similar structure when the pairwise sequence identity is high (>40% for long alignments).\n",
    "                opts = \"--min-seq-id 0.4 --cov-mode 1\"\n",
    "                cmd = \"%s easy-cluster %s %s %s %s\" % (mmseqs_path, fasta_file, prefix, prefix, opts)\n",
    "                process = subprocess.run(cmd, shell=True, check=True)\n",
    "                if os.path.isdir(prefix):\n",
    "                    shutil.rmtree(prefix)\n",
    "\n",
    "            # For each line...\n",
    "            for line in Jglobals.parse_file(cluster_clusters):\n",
    "\n",
    "                # Get cluster\n",
    "                cluster_id, unientry = line.split(\"\\t\")\n",
    "                if cluster_id not in cluster_ids:\n",
    "                    cluster_num += 1\n",
    "                    cluster_ids.add(cluster_id)\n",
    "                if unientry not in clusters:\n",
    "                    clusters.setdefault(unientry, cluster_num)\n",
    "\n",
    "        # Write\n",
    "        Jglobals.write(json_file, json.dumps(clusters, sort_keys=True, indent=4, separators=(\",\", \": \")))\n",
    "\n",
    "    else:\n",
    "        with open(json_file) as f:\n",
    "            clusters = json.load(f)\n",
    "\n",
    "    # For each TF...\n",
    "    for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "        if tf.species != s:\n",
    "                continue\n",
    "        tf.cluster_num = clusters[tf.unientry]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------#\n",
    "# Output TSV  #\n",
    "#-------------#\n",
    "\n",
    "# Initialize\n",
    "fields = [\"Gene Name\", \"Gene ID\", \"Species\", \"UniProt Accession\", \"UniProt Entry\", \"Status\", \"Sequence\", \"Family\", \"Cluster\",\n",
    "          \"Evidence\", \"JASPAR\", \"HOCOMOCO\", \"ChIP-Atlas\", \"CistromeDB\", \"GTRD\", \"ReMap\", \"DAP-seq\", \"HT-SELEX\",\n",
    "          \"CIS-BP\", \"UniPROBE\", \"SMiLE-seq\"]\n",
    "\n",
    "# For each species...\n",
    "for s in sorted(species):\n",
    "\n",
    "    # Initialize\n",
    "    annotations_file = \"./Annotations/%s.tsv\" % species[s]\n",
    "\n",
    "    if os.path.exists(annotations_file):\n",
    "        os.remove(annotations_file)\n",
    "\n",
    "    # Write\n",
    "    Jglobals.write(annotations_file, \"\\t\".join(fields))\n",
    "    for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "        if tf.species != s:\n",
    "            continue\n",
    "        Jglobals.write(annotations_file, tf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['sna', 'Drosophila melanogaster', 3, 1, 2]\n",
      "['JUN', 'Homo sapiens', 4, 1, 3]\n",
      "['PAX7', 'Homo sapiens', 4, 1, 3]\n",
      "...\n",
      "//\n",
      "Drosophila melanogaster 1\n",
      "Homo sapiens 62\n",
      "Mus musculus 32\n",
      "//\n",
      "C2H2 ZF 15\n",
      "C2H2 ZF,MADF 1\n",
      "CUT,Homeodomain 1\n",
      "DM 1\n",
      "E2F 1\n",
      "Ets 4\n",
      "Forkhead 7\n",
      "GATA 2\n",
      "Homeodomain 15\n",
      "Homeodomain,POU 1\n",
      "Homeodomain,Paired box 2\n",
      "Nuclear receptor 13\n",
      "RFX 2\n",
      "Rel 3\n",
      "SAND 1\n",
      "Sox 10\n",
      "bHLH 9\n",
      "bZIP 7\n"
     ]
    }
   ],
   "source": [
    "#-------------#\n",
    "# Triads      #\n",
    "#-------------#\n",
    "\n",
    "# Initialize\n",
    "count = 0\n",
    "done = set()\n",
    "evidence = {}\n",
    "families = {}\n",
    "strings = []\n",
    "\n",
    "# For each species...\n",
    "for s in sorted(species):\n",
    "\n",
    "    # Initialize\n",
    "    triads_file = \"./Annotations/%s.triads.tsv\" % species[s]\n",
    "\n",
    "    if os.path.exists(triads_file):\n",
    "        os.remove(triads_file)\n",
    "\n",
    "    # Write\n",
    "    Jglobals.write(triads_file, \"\\t\".join(fields))\n",
    "    for tf in sorted(tfs, key=lambda x: x.evidence, reverse=True):\n",
    "        if tf.species != s:\n",
    "            continue\n",
    "        if tuple([tf.species, tf.cluster_num]) in done:\n",
    "            continue\n",
    "        if tf.invivo > 0 and tf.invitro > 1:\n",
    "            Jglobals.write(triads_file, tf)\n",
    "            strings.append([tf.gene_name, tf.species, tf.evidence, tf.invivo, tf.invitro])\n",
    "            evidence.setdefault(tf.species, 0)\n",
    "            evidence[tf.species] += 1\n",
    "            families.setdefault(tf.family, 0)\n",
    "            families[tf.family] += 1\n",
    "            done.add(tuple([tf.species, tf.cluster_num]))\n",
    "\n",
    "for s in strings:\n",
    "    if count < 3:\n",
    "        print(s)\n",
    "    elif count == 3:\n",
    "        print(\"...\")\n",
    "    else:\n",
    "        pass\n",
    "    count += 1\n",
    "print(\"//\")\n",
    "for e in sorted(evidence):\n",
    "    print(e, evidence[e])\n",
    "print(\"//\")\n",
    "for f in sorted(families):\n",
    "    print(f, families[f])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SPL1', 'Arabidopsis thaliana', 2, 1, 1]\n",
      "['SPL11', 'Arabidopsis thaliana', 2, 1, 1]\n",
      "['SPL14', 'Arabidopsis thaliana', 2, 1, 1]\n",
      "...\n",
      "//\n",
      "Arabidopsis thaliana 111\n",
      "Caenorhabditis elegans 30\n",
      "Drosophila melanogaster 40\n",
      "Homo sapiens 294\n",
      "Mus musculus 153\n",
      "Saccharomyces cerevisiae 33\n",
      "//\n",
      "ABF1 1\n",
      "AP2 22\n",
      "AP2,B3 1\n",
      "APSES 1\n",
      "ARID/BRIGHT 1\n",
      "AT hook 2\n",
      "BED ZF 3\n",
      "Brinker 1\n",
      "C2H2 ZF 105\n",
      "C2H2 ZF,Homeodomain 3\n",
      "C2H2 ZF,MADF 1\n",
      "C2HC ZF 1\n",
      "CSD 1\n",
      "CSL 1\n",
      "CUT,Homeodomain 3\n",
      "CxxC 4\n",
      "DM 3\n",
      "Dof 2\n",
      "E2F 6\n",
      "EBF1 2\n",
      "EIN3 2\n",
      "Ets 26\n",
      "Forkhead 24\n",
      "GATA 8\n",
      "GCM 1\n",
      "GCR1 1\n",
      "Grainyhead 2\n",
      "HSF 5\n",
      "Homeodomain 107\n",
      "Homeodomain,POU 10\n",
      "Homeodomain,Paired box 5\n",
      "IRF 11\n",
      "LOB 1\n",
      "MADF 1\n",
      "MADS box 4\n",
      "MBD 1\n",
      "Myb/SANT 31\n",
      "NAC/NAM 6\n",
      "Nuclear receptor 35\n",
      "POU 1\n",
      "Paired box 1\n",
      "Pipsqueak 1\n",
      "Prospero 2\n",
      "RFX 4\n",
      "Rap1 1\n",
      "Rel 4\n",
      "Runt 2\n",
      "SAND 2\n",
      "SBP 5\n",
      "SMAD 4\n",
      "Sox 20\n",
      "T-box 7\n",
      "TBP 3\n",
      "TCP 6\n",
      "TCR/CxC 1\n",
      "TEA 1\n",
      "Unknown 3\n",
      "WRKY 13\n",
      "Zinc cluster 12\n",
      "bHLH 63\n",
      "bHLH,T-box 1\n",
      "bZIP 56\n",
      "bZIP,C2H2 ZF 1\n",
      "p53 2\n"
     ]
    }
   ],
   "source": [
    "#-------------#\n",
    "# Duos        #\n",
    "#-------------#\n",
    "\n",
    "# Initialize\n",
    "count = 0\n",
    "done = set()\n",
    "evidence = {}\n",
    "families = {}\n",
    "strings = []\n",
    "\n",
    "# For each species...\n",
    "for s in sorted(species):\n",
    "\n",
    "    # Initialize\n",
    "    triads_file = \"./Annotations/%s.duos.tsv\" % species[s]\n",
    "\n",
    "    if os.path.exists(triads_file):\n",
    "        os.remove(triads_file)\n",
    "\n",
    "    # Write\n",
    "    Jglobals.write(triads_file, \"\\t\".join(fields))\n",
    "    for tf in sorted(tfs, key=lambda x: x.evidence, reverse=True):\n",
    "        if tf.species != s:\n",
    "            continue\n",
    "        if tuple([tf.species, tf.cluster_num]) in done:\n",
    "            continue\n",
    "        if tf.invivo > 0 and tf.invitro > 0:\n",
    "            Jglobals.write(triads_file, tf)\n",
    "            strings.append([tf.gene_name, tf.species, tf.evidence, tf.invivo, tf.invitro])\n",
    "            evidence.setdefault(tf.species, 0)\n",
    "            evidence[tf.species] += 1\n",
    "            families.setdefault(tf.family, 0)\n",
    "            families[tf.family] += 1\n",
    "            done.add(tuple([tf.species, tf.cluster_num]))\n",
    "\n",
    "for s in strings:\n",
    "    if count < 3:\n",
    "        print(s)\n",
    "    elif count == 3:\n",
    "        print(\"...\")\n",
    "    else:\n",
    "        pass\n",
    "    count += 1\n",
    "print(\"//\")\n",
    "for e in sorted(evidence):\n",
    "    print(e, evidence[e])\n",
    "print(\"//\")\n",
    "for f in sorted(families):\n",
    "    print(f, families[f])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SPL1', 'Arabidopsis thaliana', 2, 1, 1]\n",
      "['SPL11', 'Arabidopsis thaliana', 2, 1, 1]\n",
      "['SPL14', 'Arabidopsis thaliana', 2, 1, 1]\n",
      "...\n",
      "//\n",
      "Arabidopsis thaliana 111\n",
      "Caenorhabditis elegans 30\n",
      "Drosophila melanogaster 41\n",
      "Homo sapiens 312\n",
      "Mus musculus 174\n",
      "Saccharomyces cerevisiae 33\n",
      "//\n",
      "ABF1 1\n",
      "AP2 22\n",
      "AP2,B3 1\n",
      "APSES 1\n",
      "ARID/BRIGHT 1\n",
      "AT hook 2\n",
      "BED ZF 3\n",
      "Brinker 1\n",
      "C2H2 ZF 107\n",
      "C2H2 ZF,Homeodomain 3\n",
      "C2H2 ZF,MADF 1\n",
      "C2HC ZF 1\n",
      "CSD 1\n",
      "CSL 1\n",
      "CUT,Homeodomain 3\n",
      "CxxC 4\n",
      "DM 5\n",
      "Dof 2\n",
      "E2F 6\n",
      "EBF1 2\n",
      "EIN3 2\n",
      "Ets 27\n",
      "Forkhead 26\n",
      "GATA 8\n",
      "GCM 2\n",
      "GCR1 1\n",
      "Grainyhead 2\n",
      "HSF 5\n",
      "Homeodomain 133\n",
      "Homeodomain,POU 12\n",
      "Homeodomain,Paired box 5\n",
      "IRF 11\n",
      "LOB 1\n",
      "MADF 1\n",
      "MADS box 4\n",
      "MBD 1\n",
      "Myb/SANT 31\n",
      "NAC/NAM 6\n",
      "Nuclear receptor 35\n",
      "POU 1\n",
      "Paired box 1\n",
      "Pipsqueak 1\n",
      "Prospero 2\n",
      "RFX 4\n",
      "Rap1 1\n",
      "Rel 4\n",
      "Runt 2\n",
      "SAND 2\n",
      "SBP 5\n",
      "SMAD 4\n",
      "Sox 21\n",
      "T-box 7\n",
      "TBP 3\n",
      "TCP 6\n",
      "TCR/CxC 1\n",
      "TEA 1\n",
      "Unknown 3\n",
      "WRKY 13\n",
      "Zinc cluster 12\n",
      "bHLH 66\n",
      "bHLH,T-box 1\n",
      "bZIP 56\n",
      "bZIP,C2H2 ZF 1\n",
      "p53 2\n"
     ]
    }
   ],
   "source": [
    "#-------------#\n",
    "# Evidence 2+ #\n",
    "#-------------#\n",
    "\n",
    "# Initialize\n",
    "count = 0\n",
    "done = set()\n",
    "evidence = {}\n",
    "families = {}\n",
    "strings = []\n",
    "\n",
    "# For each species...\n",
    "for s in sorted(species):\n",
    "\n",
    "    # Initialize\n",
    "    triads_file = \"./Annotations/%s.multiple_evidence.tsv\" % species[s]\n",
    "\n",
    "    if os.path.exists(triads_file):\n",
    "        os.remove(triads_file)\n",
    "\n",
    "    # Write\n",
    "    Jglobals.write(triads_file, \"\\t\".join(fields))\n",
    "    for tf in sorted(tfs, key=lambda x: x.evidence, reverse=True):\n",
    "        if tf.species != s:\n",
    "            continue\n",
    "        if tuple([tf.species, tf.cluster_num]) in done:\n",
    "            continue\n",
    "        if tf.evidence > 1:\n",
    "            Jglobals.write(triads_file, tf)\n",
    "            strings.append([tf.gene_name, tf.species, tf.evidence, tf.invivo, tf.invitro])\n",
    "            evidence.setdefault(tf.species, 0)\n",
    "            evidence[tf.species] += 1\n",
    "            families.setdefault(tf.family, 0)\n",
    "            families[tf.family] += 1\n",
    "            done.add(tuple([tf.species, tf.cluster_num]))\n",
    "\n",
    "for s in strings:\n",
    "    if count < 3:\n",
    "        print(s)\n",
    "    elif count == 3:\n",
    "        print(\"...\")\n",
    "    else:\n",
    "        pass\n",
    "    count += 1\n",
    "print(\"//\")\n",
    "for e in sorted(evidence):\n",
    "    print(e, evidence[e])\n",
    "print(\"//\")\n",
    "for f in sorted(families):\n",
    "    print(f, families[f])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Intersection with list of ~400 putative TFs\n",
    "# r.lovering@ucl.ac.uk\n",
    "\n",
    "# Initialize\n",
    "gene_names = set()\n",
    "tf_file = \"./Annotations/Rende_400TFs_checked.tsv\"\n",
    "\n",
    "if os.path.exists(tf_file):\n",
    "    os.remove(tf_file)\n",
    "\n",
    "# Write\n",
    "Jglobals.write(tf_file, \"\\t\".join(fields))\n",
    "\n",
    "# For each line...\n",
    "for line in Jglobals.parse_tsv_file(\"./Data/Rende_400TFs_checked.txt\"):\n",
    "    if line[3] is None:\n",
    "        continue\n",
    "    gene_names.add(line[5])\n",
    "\n",
    "# For each TF...\n",
    "for tf in sorted(tfs, key=lambda x: x.evidence, reverse=True):\n",
    "    if tf.species != \"Homo sapiens\":\n",
    "        continue\n",
    "    if tf.gene_name in gene_names:\n",
    "        Jglobals.write(tf_file, tf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------#\n",
    "# Extra stuff #\n",
    "#-------------#\n",
    "\n",
    "# # Get orthoDB cluster\n",
    "# codec = coreapi.codecs.CoreJSONCodec()\n",
    "# for uniacc in tf._uniaccs:\n",
    "#     json_file = os.path.join(args.orthodb, \"%s.json\" % uniacc)\n",
    "#     if not os.path.exists(json_file):\n",
    "#         client = coreapi.Client()\n",
    "#         response = client.get(\n",
    "#             \"https://www.orthodb.org/search?query=%s&level=2759&species=2759\" % uniacc)\n",
    "#         json_obj = json.loads(codec.encode(response))\n",
    "#         with open(json_file, \"w\") as j:\n",
    "#             j.write(json.dumps(json_obj, sort_keys=True, indent=4, separators=(\",\", \": \")))\n",
    "#     with open(json_file, \"r\") as j:  \n",
    "#         json_obj = json.load(j)\n",
    "#         for orthodb in json_obj[\"data\"]:\n",
    "#             tf.orthodb.add(orthodb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #-------------#\n",
    "# # Cluster     #\n",
    "# # (CIS-BP)    #\n",
    "# #-------------#\n",
    "\n",
    "# # The following code is adapted from:\n",
    "# # https://github.com/wassermanlab/JASPAR-profile-inference/blob/master/finfer_profile.py\n",
    "\n",
    "# import math\n",
    "# import numpy as np\n",
    "# from infer_profile import _filter_results_by_Rost, _get_pid\n",
    "\n",
    "# def get_members(tf):\n",
    "\n",
    "#     # Initialize\n",
    "#     members = []\n",
    "#     tf_DBDs = [dbd[0] for dbd in pfams[tf.unientry]]\n",
    "#     tf_alignments = [dbd[1] for dbd in pfams[tf.unientry]]\n",
    "#     seq = Seq(tf.sequence, IUPAC.protein)\n",
    "#     seq_record = SeqRecord(seq, id=tf.unientry, name=tf.unientry, description=tf.unientry)\n",
    "\n",
    "#     # Get cut-offs on the percentage of sequence identity\n",
    "#     cutoffs = {}\n",
    "#     for pfam_ac in pfam_cutoffs:\n",
    "#         if pfam_cutoffs[pfam_ac][0] in tf_DBDs:\n",
    "#             cutoffs.setdefault(pfam_cutoffs[pfam_ac][0], pfam_cutoffs[pfam_ac][1])\n",
    "\n",
    "#     # BLAST+ search\n",
    "#     blast_results = BLAST(seq_record)\n",
    "\n",
    "#     # Filter results\n",
    "#     for filtered_result in sorted(_filter_results_by_Rost(blast_results), key=lambda x: x[5], reverse=True):\n",
    "\n",
    "#         # Both TFs have same DBD composition\n",
    "#         if [dbd[0] for dbd in pfams[filtered_result[1]]] == tf_DBDs:\n",
    "            \n",
    "#             # Inference: percentage of sequence identity\n",
    "#             if tf_DBDs:\n",
    "\n",
    "#                 skip = False\n",
    "#                 alignments = [dbd[1] for dbd in pfams[filtered_result[1]]]\n",
    "#                 for a in range(len(alignments)):\n",
    "#                     pid = _get_pid(tf_alignments[a], alignments[a])\n",
    "#                     if pid < cutoffs[tf_DBDs[a]]:\n",
    "#                         skip = True\n",
    "#                         break\n",
    "#                 if not skip:\n",
    "#                     members.append(filtered_result[1])\n",
    "\n",
    "#             # From PMID:25215497;\n",
    "#             # For the remaining classes, with suggestive but insufficient data, we chose a threshold of 70%,\n",
    "#             # which is the mean, median, and mode threshold across all DBD classes.\n",
    "#             else:\n",
    "#                 if filtered_result[6] < 70:\n",
    "#                     continue\n",
    "#                 members.append(filtered_result[1])\n",
    "\n",
    "#     return(members)\n",
    "\n",
    "# def BLAST(seq_record):\n",
    "\n",
    "#     # Initialize\n",
    "#     blast_results = set()\n",
    "#     outfmt = \"sseqid pident length qstart qend sstart send evalue bitscore ppos qlen slen\"\n",
    "\n",
    "#     # Run BLAST+\n",
    "#     cmd = \"blastp -db %s -outfmt \\\"6 %s\\\"\" % (fasta_file, outfmt)\n",
    "#     process = subprocess.Popen([cmd], shell=True, stdin=subprocess.PIPE, stdout=subprocess.PIPE)\n",
    "#     fasta_sequence = \">%s\\n%s\" % (seq_record.id, seq_record.seq)\n",
    "#     process.stdin.write(fasta_sequence.encode())\n",
    "#     (blast_records, blast_errors) = process.communicate()\n",
    "\n",
    "#     # For each BLAST+ record...\n",
    "#     for blast_record in blast_records.decode(\"utf-8\").split(\"\\n\"):\n",
    "\n",
    "#         # Custom BLAST+ record:\n",
    "#         # (1) identifier of target sequence;\n",
    "#         # (2) percentage of identical matches;\n",
    "#         # (3) alignment length;\n",
    "#         # (4-5, 6-7) start and end-position in query and in target;\n",
    "#         # (8) E-value;\n",
    "#         # (9) bit score;\n",
    "#         # (10) percentage of positive-scoring matches; and\n",
    "#         # (4-7, 11, 12) joint coverage (i.e. square root of the coverage\n",
    "#         # on the query and the target).\n",
    "#         blast_record = blast_record.split(\"\\t\")\n",
    "\n",
    "#         # Skip if not a BLAST+ record\n",
    "#         if len(blast_record) != 12: continue\n",
    "\n",
    "#         # Get BLAST+ record\n",
    "#         target_id = blast_record[0]\n",
    "#         percent_identities = float(blast_record[1])\n",
    "#         alignment_length = int(blast_record[2])\n",
    "#         query_start_end = \"%s-%s\" % (blast_record[3], blast_record[4])\n",
    "#         target_start_end = \"%s-%s\" % (blast_record[5], blast_record[6])\n",
    "#         e_value = float(blast_record[7])\n",
    "#         score = float(blast_record[8])\n",
    "#         percent_similarity = float(blast_record[9])\n",
    "#         query_aligned_residues = int(blast_record[4]) - int(blast_record[3]) + 1\n",
    "#         query_length = float(blast_record[10])\n",
    "#         target_aligned_residues = int(blast_record[6]) - int(blast_record[5]) + 1\n",
    "#         target_length = float(blast_record[11])\n",
    "#         query_coverage = query_aligned_residues * 100 / query_length\n",
    "#         target_coverage = target_aligned_residues * 100 / target_length\n",
    "#         joint_coverage = math.sqrt(query_coverage * target_coverage)\n",
    "\n",
    "#         # Add BLAST+ record to search results\n",
    "#         blast_results.add((seq_record.id, target_id, query_start_end, target_start_end, e_value, score, percent_identities, alignment_length, percent_similarity, joint_coverage))\n",
    "\n",
    "#     # Return results sorted by score\n",
    "#     return(list(sorted(blast_results, key=lambda x: x[-1], reverse=True)))\n",
    "\n",
    "# # Initialize\n",
    "# with open(os.path.join(jaspar_dir, \"files\", \"pfam-DBDs.json\")) as f:\n",
    "#     pfam_cutoffs = json.load(f)\n",
    "\n",
    "# # Get species\n",
    "# species = {}\n",
    "# for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "#     species.setdefault(tf.species, tf.species.replace(\" \", \"_\"))\n",
    "\n",
    "# # For each species...\n",
    "# for s in sorted(species):\n",
    "\n",
    "#     # Initialize\n",
    "#     cluster_num = 0\n",
    "#     clusters = {}\n",
    "#     clusters_dir = \"./Data/Clusters/%s\" % species[s]\n",
    "#     family_num = 0\n",
    "#     families = {}\n",
    "\n",
    "#     if not os.path.isdir(clusters_dir):\n",
    "#         os.mkdir(clusters_dir)\n",
    "\n",
    "#     # Get families\n",
    "#     for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "#         if tf.species != s:\n",
    "#             continue\n",
    "#         if tf.family not in families:\n",
    "#             family_num += 1\n",
    "#             families.setdefault(tf.family, family_num)\n",
    "#             prefix = os.path.join(clusters_dir, str(families[tf.family]))\n",
    "#             fasta_file = \"%s.fasta\" % prefix\n",
    "#             if os.path.exists(fasta_file):\n",
    "#                 os.remove(fasta_file)\n",
    "\n",
    "#     # Create family-specific FASTA files\n",
    "#     for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "#         if tf.species != s:\n",
    "#             continue\n",
    "#         prefix = os.path.join(clusters_dir, str(families[tf.family]))\n",
    "#         fasta_file = \"%s.fasta\" % prefix\n",
    "#         Jglobals.write(fasta_file, \">%s\\n%s\" % (tf.unientry, tf.sequence))\n",
    "\n",
    "#     # Skip if JSON file already exists\n",
    "#     json_file = \"%s/%s.json\" % (clusters_dir, species[s])\n",
    "#     if not os.path.exists(json_file):\n",
    "\n",
    "#         for family in sorted(families.values()):\n",
    "\n",
    "#             # Make BLAST+ databases\n",
    "#             prefix = os.path.join(clusters_dir, str(family))\n",
    "#             fasta_file = \"%s.fasta\" % prefix\n",
    "#             if not os.path.exists(\"%s.psq\" % fasta_file):\n",
    "#                 cmd = \"makeblastdb -in %s -dbtype prot\" % fasta_file\n",
    "#                 process = subprocess.run([cmd], shell=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)\n",
    "\n",
    "#             for tf in sorted(tfs, key=lambda x: x.evidence, reverse=True):\n",
    "\n",
    "#                 # Exit loop\n",
    "#                 if tf.evidence == 0:\n",
    "#                     break\n",
    "#                 if tf.species != s:\n",
    "#                     continue\n",
    "#                 if families[tf.family] != family:\n",
    "#                     continue\n",
    "#                 if tf.unientry in clusters:\n",
    "#                     continue\n",
    "\n",
    "#                 # Get cluster members\n",
    "#                 members = get_members(tf)\n",
    "#                 if members:\n",
    "#                     cluster_num += 1\n",
    "#                     for member in members:\n",
    "#                         if member in clusters:\n",
    "#                             continue\n",
    "#                         clusters.setdefault(member, cluster_num)\n",
    "\n",
    "#         # Write\n",
    "#         Jglobals.write(json_file, json.dumps(clusters, sort_keys=True, indent=4, separators=(\",\", \": \")))\n",
    "\n",
    "# # with open(json_file) as f:\n",
    "# #     clusters = json.load(f)\n",
    "\n",
    "# # for tf in sorted(tfs, key=lambda x: x.gene_name):\n",
    "# #     if tf.uniacc in clusters:\n",
    "# #         tf.cluster_num = clusters[tf.uniacc]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
